%% Generated by Sphinx.
\def\sphinxdocclass{report}
\documentclass[letterpaper,10pt,english]{sphinxmanual}
\ifdefined\pdfpxdimen
   \let\sphinxpxdimen\pdfpxdimen\else\newdimen\sphinxpxdimen
\fi \sphinxpxdimen=.75bp\relax

\usepackage[utf8]{inputenc}
\ifdefined\DeclareUnicodeCharacter
 \ifdefined\DeclareUnicodeCharacterAsOptional\else
  \DeclareUnicodeCharacter{00A0}{\nobreakspace}
\fi\fi
\usepackage{cmap}
\usepackage[T1]{fontenc}
\usepackage{amsmath,amssymb,amstext}
\usepackage{babel}
\usepackage{times}
\usepackage[Bjarne]{fncychap}
\usepackage{longtable}
\usepackage{sphinx}

\usepackage{geometry}
\usepackage{multirow}
\usepackage{eqparbox}

% Include hyperref last.
\usepackage{hyperref}
% Fix anchor placement for figures with captions.
\usepackage{hypcap}% it must be loaded after hyperref.
% Set up styles of URL: it should be placed after hyperref.
\urlstyle{same}

\addto\captionsenglish{\renewcommand{\figurename}{Fig.}}
\addto\captionsenglish{\renewcommand{\tablename}{Table}}
\addto\captionsenglish{\renewcommand{\literalblockname}{Listing}}

\addto\extrasenglish{\def\pageautorefname{page}}

\setcounter{tocdepth}{1}



\title{CrossMap Documentation}
\date{Nov 06, 2018}
\release{0.2.6}
\author{Liguo Wang}
\newcommand{\sphinxlogo}{}
\renewcommand{\releasename}{Release}
\makeindex

\begin{document}

\maketitle
\sphinxtableofcontents
\phantomsection\label{\detokenize{index::doc}}
\noindent\scalebox{0.500000}{\sphinxincludegraphics[width=750\sphinxpxdimen,height=150\sphinxpxdimen]{{logo}.png}}


\begin{itemize}
\item {} 
CrossMap is a program for convenient conversion of genome coordinates (or annotation files)
between \sphinxstyleemphasis{different assemblies} (such as Human \sphinxhref{http://www.ncbi.nlm.nih.gov/assembly/2928/}{hg18 (NCBI36)}
\textless{}\textgreater{} \sphinxhref{http://www.ncbi.nlm.nih.gov/assembly/2758/}{hg19 (GRCh37)}, Mouse \sphinxhref{http://www.ncbi.nlm.nih.gov/assembly/165668/}{mm9 (MGSCv37)}
\textless{}\textgreater{} \sphinxhref{http://www.ncbi.nlm.nih.gov/assembly/327618/}{mm10 (GRCm38)}).

\item {} 
It supports most commonly used file formats including SAM/BAM, Wiggle/BigWig, BED, GFF/GTF, VCF.

\item {} 
CrossMap is designed to liftover genome coordinates between assemblies. It's \sphinxstyleemphasis{not} a program
for aligning sequences to reference genome.

\item {} 
We \sphinxstyleemphasis{do not} recommend using CrossMap to convert genome coordinates between species.

\end{itemize}


\chapter{Why CrossMap ?}
\label{\detokenize{index:what-is-crossmap}}\label{\detokenize{index:why-crossmap}}
Full genome sequencing, especially mammalian (eg. human) genomes, requires extensive, continuous
efforts. Therefore reference genome assemblies are subject to change and refinement from time
to time. Generally, researchers need to convert results that have been analyzed according to
old assemblies to newer versions or \sphinxstyleemphasis{vice versa},  to facilitate meta-analysis, direct comparison
as well as data integration and visualization.

Several useful conversion tools have been developed:
\begin{itemize}
\item {} 
\sphinxhref{http://genome.ucsc.edu/cgi-bin/hgLiftOver}{UCSC liftover tool} only supports BED input.

\item {} 
\sphinxhref{http://www.ncbi.nlm.nih.gov/genome/tools/remap}{NCBI remap} support BED, GFF, GTF, VCF, etc

\item {} 
\sphinxhref{https://usegalaxy.org/}{Galaxy} (Based on UCSC liftover tool) supports BED, GFF, GTF input.

\item {} 
\sphinxhref{http://www.ensembl.org/Homo\_sapiens/Tools/AssemblyConverter?db=core}{Ensembl assembly converter}
supports BED, GFF, GTF, PSL input, but output is GFF only. (\sphinxstylestrong{Update:} The original \sphinxquotedblleft{}assembly converter\sphinxquotedblright{} has been retired. Starting from 2015, Ensembl uses \sphinxhref{http://www.ensembl.org/Homo\_sapiens/Tools/AssemblyConverter?db=core}{CrossMap} to perform genome coordinate conversion.)

\item {} 
\sphinxhref{https://pypi.python.org/pypi/pyliftover}{pyliftover} \sphinxquotedblleft{}only does conversion of point
coordinates, that is, unlike liftOver, it does not convert ranges, nor does it provide any
special facilities to work with BED files\sphinxquotedblright{}.

\end{itemize}

But none have the functionality to convert files in BAM/SAM or BigWig format. This is a significant
gap in computational genomics tools, since these formats are the ones most widely used
for representing high-throughput sequencing data such as RNA-seq, ChIP-seq, DNA-seq, etc.


\chapter{Who is using CrossMap ?}
\label{\detokenize{index:who-is-using-crossmap}}\begin{itemize}
\item {} 
\sphinxhref{http://www.ensembl.org/Homo\_sapiens/Tools/AssemblyConverter?db=core}{Ensembl}

\item {} 
\sphinxhref{https://basespace.illumina.com/apps/}{Illumina BaseSpace}

\item {} 
\sphinxhref{http://bcbio-nextgen.readthedocs.org/en/latest/contents/introduction.html}{bcbio-nextgen}

\item {} 
\sphinxhref{https://hpc.nih.gov/apps/crossmap.html}{NIH Biowulf \& Helix}

\item {} 
\sphinxhref{https://wiki.gacrc.uga.edu/wiki/CrossMap}{Georgia Advanced Computing Resource Center}

\item {} 
\sphinxhref{https://www.gnu.org/software/guix/packages/}{GNU Guix}

\end{itemize}


\chapter{How CrossMap works?}
\label{\detokenize{index:how-crossmap-works}}
\noindent\scalebox{0.850000}{\sphinxincludegraphics[width=600\sphinxpxdimen,height=250\sphinxpxdimen]{{howitworks}.png}}


\section{Algorithm}
\label{\detokenize{index:algorithm}}
CrossMap first determines the correspondence between genome assemblies from
UCSC \sphinxhref{http://genome.ucsc.edu/goldenPath/help/chain.html}{chain} file (chain file
describes the pair-wise alignments between two genomes). Genome intervals will be stored in
\sphinxhref{http://en.wikipedia.org/wiki/Interval\_tree}{interval tree} data structure,
which  allows one to efficiently find all intervals that overlap with any given interval or point.
Then CrossMap remaps each entry in BAM/SAM, BED, GFF/GTF, VCF file to the target assembly by querying the \sphinxhref{http://en.wikipedia.org/wiki/Interval\_tree}{interval tree}.
Exon/intron structure in BED file; spliced alignments, paired alignments, insert size, header
section, SAM flags in BAM/SAM file; reference alleles, indels in VCF file will be processed
properly.

For Wiggle/BigWig format files, line-by-line computation will be very slow. To increase speed,
CrossMap groups consecutive coordinates with the same coverage score into bins (i.e. genomic regions),
then remaps those regions one-by-one to the target assembly by querying the interval tree.  In other words, Wiggle/BigWig files will
be converted into \sphinxhref{http://genome.ucsc.edu/goldenPath/help/bedgraph.html}{bedGraph} format
internally, which will be converted into BigWig format (if UCSCâ€™s `\sphinxhref{http://hgdownload.cse.ucsc.edu/admin/exe/}{wigToBigWig}`
executable exists and is callable).


\section{Time complexity}
\label{\detokenize{index:time-complexity}}
Assume there are N lines in the chain file. CrossMap loads the chain file first and process
the query file line by line. Thus the space complexity is O(N). For each query region (s,t),
it takes O(logN) time to locate which chain(s) overlap with s and t. Then it takes O(logN)
time to search the sorted ungapped alignments in this chain that overlap with s and t and
calculate the converted values for s and t in the target assembly. So in total it takes O(logN)
time to convert one query. The time complexity is O(logN*M) to convert M queries.

In practical, the time CrossMap takes increases linearly to the size of input file.


\chapter{Release history}
\label{\detokenize{index:release-history}}\begin{itemize}
\item {} 
11/05/18: Release version 0.3.0:

\end{itemize}
\begin{itemize}
\item {} 
v0.3.0 or newer will Support Python3. Previous versions support Python2.7.*

\item {} 
add \sphinxhref{https://github.com/deeptools/pyBigWig}{pyBigWig} as dependency.

\end{itemize}
\begin{itemize}
\item {} 
09/06/17: Release version 0.2.8:

\end{itemize}
\begin{itemize}
\item {} 
In Bam file lift over: fixed the bug \sphinxquotedblleft{}CrossMap does not set the unmapped read flag for the first read in pair when it is unmapped\sphinxquotedblright{}.

\item {} 
In VCF file lift over: Update the \sphinxquotedblleft{}contig field\sphinxquotedblright{} in VCF header section. Contig name and size will be changed from old assembly to new assembly.

\end{itemize}
\begin{itemize}
\item {} 
09/06/17: Release version 0.2.7:

\end{itemize}
\begin{itemize}
\item {} 
In VCF file lift over: fixed the bug \sphinxquotedblleft{}non-standard chromosome IDs were not converted\sphinxquotedblright{}.

\end{itemize}
\begin{itemize}
\item {} 
05/09/17: Release version 0.2.6:

\end{itemize}
\begin{itemize}
\item {} 
In BAM file lift over: fixed bugs during BAM file sorting and indexing steps (works with pysam v0.11.1).

\item {} 
In BAM file lift over: fixed bugs \sphinxquotedblleft{}the read group type is automatically and wrongly changed from Z to A\sphinxquotedblright{} (\sphinxurl{https://github.com/pysam-developers/pysam/issues/113}).

\end{itemize}
\begin{itemize}
\item {} 
10/7/16: Release version 0.2.5:

\end{itemize}
\begin{itemize}
\item {} 
fixed bugs during single-end BAM file conversion.

\item {} 
Add optional tags to the output BAM file. Details see: {\hyperref[\detokenize{index:bam-conversion}]{\sphinxcrossref{\DUrole{std,std-ref}{Convert BAM/SAM format files}}}}.

\end{itemize}
\begin{itemize}
\item {} 
08/18/16: Release version 0.2.4:

\end{itemize}
\begin{quote}
\begin{itemize}
\item {} 
fixed bugs during BAM file conversion:

\end{itemize}
\begin{itemize}
\item {} 
When the strand of read changes, the seq filed is reverse complemented and the quality field is reversed.

\end{itemize}
\begin{itemize}
\item {} 
In the output VCF file, if the reference allele field is empty:

\end{itemize}
\begin{itemize}
\item {} 
Use CrossMap v0.2.4. Update pysam to the latest version. And make sure chromosome IDs in the reference genome file are in the form of \sphinxquotedblleft{}chr1\sphinxquotedblright{}, \sphinxquotedblleft{}chr2\sphinxquotedblright{}, ..., \sphinxquotedblleft{}chrX\sphinxquotedblright{},\sphinxquotedblright{}chrY\sphinxquotedblright{} (but not \sphinxquotedblleft{}1\sphinxquotedblright{}, \sphinxquotedblleft{}2\sphinxquotedblright{}, ..., \sphinxquotedblleft{}X\sphinxquotedblright{},\sphinxquotedblright{}Y\sphinxquotedblright{}, in this case, pysam cannot index your reference genome file for some unknown reasons.).

\end{itemize}
\begin{itemize}
\item {} 
to upgrade, run: \sphinxstylestrong{pip install CrossMap --upgrade}

\end{itemize}
\end{quote}
\begin{itemize}
\item {} 
04/13/16: Release version 0.2.3:

\end{itemize}
\begin{itemize}
\item {} 
Same as v0.2.2.

\item {} 
Two dependency packages bx-python and pysam do not shipped with CrossMap starting from v0.2.3 .

\item {} 
Users could install CrossMap using pip: \sphinxstylestrong{pip install CrossMap}. Note: bx-python and pysam will be installed automatically if they havenâ€™t been installed before.

\end{itemize}
\begin{itemize}
\item {} 
11/10/15: Release version 0.2.2: Generate *.unmap files (regions that cannot be unambiguously converted) when converting BED, GTF, GFF files. This version also supports genePred (bed12+8) format. (Thanks for Andrew Yates from EMBL-EBI)

\item {} 
08/26/15: Release version 0.2.1: Very minor change, same as 0.2.

\item {} 
08/11/15: Release version 0.2: Fixed the bug that CrossMap will not convert wiggle format files due to name collision with bx python.

\item {} 
07/27/15: Release version 0.1.9. For VCF file conversion in v0.1.9:

\end{itemize}
\begin{itemize}
\item {} 
CrossMap uses the indexed reference genome (target assembly) sequences rather than load the entire file into memory. Users could index their reference genome file using \sphinxstylestrong{samtools faidx} before running CrossMap, otherwise CrossMap will index it automatically the first time you run it.

\item {} 
In the output VCF file, whether the chromosome IDs contain \sphinxquotedblleft{}chr\sphinxquotedblright{} or not depends on the input format.

\end{itemize}
\begin{itemize}
\item {} 
05/15/15: Release version 0.1.8: Fixed the bug that CrossMap will output invalid VCF file when the input VCF file contains a INFO field with whitespace.

\item {} 
05/04/15: Release version 0.1.7: Address the problem that CrossMap does not convert strand in inversions when input file is BED6 or BED12 format.

\item {} 
11/06/14: Release version 0.1.6: Fixed \sphinxquotedblleft{}negative coordinates\sphinxquotedblright{} bug.

\item {} 
08/05/14: Release version 0.1.5: Support compressed (*.gz, *.Z, *.z, *.bz, *.bz2, *.bzip2) wiggle file as input.

\item {} 
05/19/14: add chain files for hg38-\textgreater{}hg19, hg19-\textgreater{}hg38, hg18-\textgreater{}hg38, hg19-\textgreater{}GRCh37, GRCh37-\textgreater{}hg19. In CrossMap v0.1.4, conversion results of BAM/SAM files can be directed to STDOUT to support piping.

\item {} 
12/12/13: CrossMap was accepted by \sphinxhref{http://bioinformatics.oxfordjournals.org/content/early/2013/12/18/bioinformatics.btt730.short?rss=1}{Bioinformatics}

\item {} 
10/23/13: CrossMap (0.1.3) was released

\end{itemize}


\chapter{Installation}
\label{\detokenize{index:installation}}

\section{Use pip to install CrossMap}
\label{\detokenize{index:use-pip-to-install-crossmap}}
\begin{sphinxVerbatim}[commandchars=\\\{\}]
\PYG{n}{pip3} \PYG{n}{install} \PYG{n}{CrossMap}  \PYG{c+c1}{\PYGZsh{}Install CrossMap supporting Python3}
\PYG{n}{pip2} \PYG{n}{install} \PYG{n}{CrossMap}  \PYG{c+c1}{\PYGZsh{}Install CrossMap supporting Python2.7.*}
\end{sphinxVerbatim}


\section{Use pip to upgrade CrossMap}
\label{\detokenize{index:use-pip-to-upgrade-crossmap}}
\begin{sphinxVerbatim}[commandchars=\\\{\}]
\PYG{n}{pip3} \PYG{n}{install} \PYG{n}{CrossMap} \PYG{o}{\PYGZhy{}}\PYG{o}{\PYGZhy{}}\PYG{n}{upgrade}        \PYG{c+c1}{\PYGZsh{}upgrade CrossMap supporting Python3}
\PYG{n}{pip2} \PYG{n}{install} \PYG{n}{CrossMap} \PYG{o}{\PYGZhy{}}\PYG{o}{\PYGZhy{}}\PYG{n}{upgrade}        \PYG{c+c1}{\PYGZsh{}upgrade CrossMap supporting Python2.7.*}
\end{sphinxVerbatim}


\section{Install CrossMap from source code}
\label{\detokenize{index:install-crossmap-from-source-code}}\begin{itemize}
\item {} 
\sphinxhref{http://sourceforge.net/projects/crossmap/files}{Source code}

\item {} 
\sphinxhref{http://sourceforge.net/projects/crossmap/files/test.hg19.zip/download}{Test datsets}

\end{itemize}

\sphinxstylestrong{Prerequisite}
\begin{itemize}
\item {} 
CrossMap (version \textless{}= 0.2.9)

\end{itemize}
\begin{enumerate}
\item {} 
\sphinxhref{http://www.python.org/getit/releases/2.7/}{python2.7.*}

\item {} 
\sphinxhref{http://numpy.scipy.org/}{numpy}

\item {} 
\sphinxhref{http://cython.org/}{cython}

\item {} 
\sphinxhref{https://pypi.python.org/pypi/pysam}{pysam}

\item {} 
\sphinxhref{https://pypi.python.org/pypi/bx-python/0.7.3}{bx-python}

\end{enumerate}
\begin{itemize}
\item {} 
CrossMap (version \textgreater{}=  0.3.0)

\end{itemize}
\begin{enumerate}
\item {} 
\sphinxhref{https://www.python.org/downloads/release/python-360/}{python3}

\item {} 
\sphinxhref{http://numpy.scipy.org/}{numpy}

\item {} 
\sphinxhref{http://cython.org/}{cython}

\item {} 
\sphinxhref{https://pypi.python.org/pypi/pysam}{pysam}

\item {} 
\sphinxhref{https://pypi.python.org/pypi/bx-python/0.7.3}{bx-python}

\item {} 
\sphinxhref{https://github.com/deeptools/pyBigWig}{pyBigWig}

\end{enumerate}

\begin{sphinxVerbatim}[commandchars=\\\{\}]
\PYGZdl{} tar zxf CrossMap\PYGZhy{}VERSION.tar.gz

\PYGZdl{} cd CrossMap\PYGZhy{}VERSION

\PYGZsh{} install CrossMap to default location. In Linux/Unix, this location is like:
\PYGZsh{} /home/user/lib/python2.7/site\PYGZhy{}packages/
\PYGZdl{} python setup.py install

\PYGZsh{} or you can install CrossMap to a specified location:
\PYGZdl{} python setup.py install \PYGZhy{}\PYGZhy{}root=/home/user/CrossMap

\PYGZsh{} setup PYTHONPATH. Skip this step if CrossMap was installed to default location.
\PYGZdl{} export PYTHONPATH=/home/user/CrossMap/usr/local/lib/python2.7/site\PYGZhy{}packages:\PYGZdl{}PYTHONPATH.

\PYGZsh{} Skip this step if CrossMap was installed to default location.
\PYGZdl{} export PATH=/home/user/CrossMap/usr/local/bin:\PYGZdl{}PATH
\end{sphinxVerbatim}

NOTE:
\begin{enumerate}
\item {} 
Mac users need to download and install \sphinxhref{https://developer.apple.com/xcode/}{Xcode}
command line tools.

\end{enumerate}


\chapter{Input and Output}
\label{\detokenize{index:input-and-output}}
CrossMap basically needs 2 input files.  \sphinxhref{http://genome.ucsc.edu/goldenPath/help/chain.html}{chain}
format file describing genom-wide pairwise alignments between assemblies and the file  containing
genome coordinates that you want to convert to different assembly. If input file is in VCF
format, a reference genome sequence file(in FASTA format) is needed.


\section{Chain file}
\label{\detokenize{index:chain-file}}
Example of \sphinxhref{http://genome.ucsc.edu/goldenPath/help/chain.html}{chain} file:

\begin{sphinxVerbatim}[commandchars=\\\{\}]
\PYG{n}{chain} \PYG{l+m+mi}{4900} \PYG{n}{chrY} \PYG{l+m+mi}{58368225} \PYG{o}{+} \PYG{l+m+mi}{25985403} \PYG{l+m+mi}{25985638} \PYG{n}{chr5} \PYG{l+m+mi}{151006098} \PYG{o}{\PYGZhy{}} \PYG{l+m+mi}{43257292} \PYG{l+m+mi}{43257528} \PYG{l+m+mi}{1}
 \PYG{l+m+mi}{9}       \PYG{l+m+mi}{1}       \PYG{l+m+mi}{0}
 \PYG{l+m+mi}{10}      \PYG{l+m+mi}{0}       \PYG{l+m+mi}{5}
 \PYG{l+m+mi}{61}      \PYG{l+m+mi}{4}       \PYG{l+m+mi}{0}
 \PYG{l+m+mi}{16}      \PYG{l+m+mi}{0}       \PYG{l+m+mi}{4}
 \PYG{l+m+mi}{42}      \PYG{l+m+mi}{3}       \PYG{l+m+mi}{0}
 \PYG{l+m+mi}{16}      \PYG{l+m+mi}{0}       \PYG{l+m+mi}{8}
 \PYG{l+m+mi}{14}      \PYG{l+m+mi}{1}       \PYG{l+m+mi}{0}
 \PYG{l+m+mi}{3}       \PYG{l+m+mi}{7}       \PYG{l+m+mi}{0}
 \PYG{l+m+mi}{48}

 \PYG{n}{chain} \PYG{l+m+mi}{4900} \PYG{n}{chrY} \PYG{l+m+mi}{58368225} \PYG{o}{+} \PYG{l+m+mi}{25985406} \PYG{l+m+mi}{25985566} \PYG{n}{chr5} \PYG{l+m+mi}{151006098} \PYG{o}{\PYGZhy{}} \PYG{l+m+mi}{43549808} \PYG{l+m+mi}{43549970} \PYG{l+m+mi}{2}
 \PYG{l+m+mi}{16}      \PYG{l+m+mi}{0}       \PYG{l+m+mi}{2}
 \PYG{l+m+mi}{60}      \PYG{l+m+mi}{4}       \PYG{l+m+mi}{0}
 \PYG{l+m+mi}{10}      \PYG{l+m+mi}{0}       \PYG{l+m+mi}{4}
 \PYG{l+m+mi}{70}
\end{sphinxVerbatim}

\sphinxstylestrong{UCSC built chain files (Human, Homo sapiens)}
\begin{itemize}
\item {} 
\sphinxhref{http://hgdownload.soe.ucsc.edu/goldenPath/hg38/liftOver/hg38ToHg19.over.chain.gz}{hg38ToHg19.over.chain.gz} (Chain file needed to convert hg38 to hg19)

\item {} 
\sphinxhref{http://hgdownload.soe.ucsc.edu/goldenPath/hg19/liftOver/hg19ToHg38.over.chain.gz}{hg19ToHg38.over.chain.gz} (Chain file needed to convert hg19 to hg38)

\item {} 
\sphinxhref{http://hgdownload.soe.ucsc.edu/goldenPath/hg18/liftOver/hg18ToHg38.over.chain.gz}{hg18ToHg38.over.chain.gz} (Chain file needed to convert hg18 to hg38)

\item {} 
\sphinxhref{http://hgdownload.soe.ucsc.edu/goldenPath/hg19/liftOver/hg19ToHg18.over.chain.gz}{hg19ToHg18.over.chain.gz} (Chain file needed to convert hg19 to hg18)

\item {} 
\sphinxhref{http://hgdownload.soe.ucsc.edu/goldenPath/hg19/liftOver/hg19ToHg17.over.chain.gz}{hg19ToHg17.over.chain.gz} (Chain file needed to convert hg19 to hg17)

\item {} 
\sphinxhref{http://hgdownload.soe.ucsc.edu/goldenPath/hg18/liftOver/hg18ToHg19.over.chain.gz}{hg18ToHg19.over.chain.gz} (Chain file needed to convert hg18 to hg19)

\item {} 
\sphinxhref{http://hgdownload.soe.ucsc.edu/goldenPath/hg18/liftOver/hg18ToHg17.over.chain.gz}{hg18ToHg17.over.chain.gz} (Chain file needed to convert hg18 to hg17)

\item {} 
\sphinxhref{http://hgdownload.soe.ucsc.edu/goldenPath/hg17/liftOver/hg17ToHg19.over.chain.gz}{hg17ToHg19.over.chain.gz} (Chain file needed to convert hg17 to hg19)

\item {} 
\sphinxhref{http://hgdownload.soe.ucsc.edu/goldenPath/hg17/liftOver/hg17ToHg18.over.chain.gz}{hg17ToHg18.over.chain.gz} (Chain file needed to convert hg17 to hg18)

\item {} 
\sphinxhref{http://sourceforge.net/projects/crossmap/files/chain\_files/GRCh37ToHg19.over.chain.gz/download}{GRCh37ToHg19.over.chain.gz} (Chain file needed to convert GRCh37 to hg19)

\item {} 
\sphinxhref{http://sourceforge.net/projects/crossmap/files/chain\_files/hg19ToGRCh37.over.chain.gz/download}{hg19ToGRCh37.over.chain.gz} (Chain file needed to convert hg19 to GRCh37)

\end{itemize}

\sphinxstylestrong{UCSC built chain files (Mouse, Mus musculus)}
\begin{itemize}
\item {} 
\sphinxhref{http://hgdownload.soe.ucsc.edu/goldenPath/mm10/liftOver/mm10ToMm9.over.chain.gz}{mm10ToMm9.over.chain.gz}  (Chain file needed to convert mm10 to mm9)

\item {} 
\sphinxhref{http://hgdownload.soe.ucsc.edu/goldenPath/mm9/liftOver/mm9ToMm10.over.chain.gz}{mm9ToMm10.over.chain.gz}  (Chain file needed to convert mm9 to mm10)

\item {} 
\sphinxhref{http://hgdownload.soe.ucsc.edu/goldenPath/mm9/liftOver/mm9ToMm8.over.chain.gz}{mm9ToMm8.over.chain.gz} (Chain file needed to convert mm9 to mm8)

\end{itemize}

\sphinxstylestrong{UCSC Chain file of other species can be downloaded from:} \sphinxurl{http://hgdownload.soe.ucsc.edu/downloads.html}

\sphinxstylestrong{Ensembl built chain files (Human, Homo sapiens)}
\begin{itemize}
\item {} 
NCBI34 \textless{}=\textgreater{} GRCh38

\end{itemize}
\begin{itemize}
\item {} 
\sphinxhref{https://sourceforge.net/projects/crossmap/files/Ensembl\_chain\_files/homo\_sapiens\%28human\%29/NCBI34\_to\_GRCh38.chain.gz/download}{NCBI34\_to\_GRCh38.chain.gz}

\item {} 
\sphinxhref{https://sourceforge.net/projects/crossmap/files/Ensembl\_chain\_files/homo\_sapiens\%28human\%29/GRCh38\_to\_NCBI34.chain.gz/download}{GRCh38\_to\_NCBI34.chain.gz}

\end{itemize}
\begin{itemize}
\item {} 
NCBI35 \textless{}=\textgreater{} GRCh38

\end{itemize}
\begin{itemize}
\item {} 
\sphinxhref{https://sourceforge.net/projects/crossmap/files/Ensembl\_chain\_files/homo\_sapiens\%28human\%29/NCBI35\_to\_GRCh38.chain.gz/download}{NCBI35\_to\_GRCh38.chain.gz}

\item {} 
\sphinxhref{https://sourceforge.net/projects/crossmap/files/Ensembl\_chain\_files/homo\_sapiens\%28human\%29/GRCh38\_to\_NCBI35.chain.gz/download}{GRCh38\_to\_NCBI35.chain.gz}

\end{itemize}
\begin{itemize}
\item {} 
NCBI36 \textless{}=\textgreater{} GRCh38

\end{itemize}
\begin{itemize}
\item {} 
\sphinxhref{https://sourceforge.net/projects/crossmap/files/Ensembl\_chain\_files/homo\_sapiens\%28human\%29/NCBI36\_to\_GRCh38.chain.gz/download}{NCBI36\_to\_GRCh38.chain.gz}

\item {} 
\sphinxhref{https://sourceforge.net/projects/crossmap/files/Ensembl\_chain\_files/homo\_sapiens\%28human\%29/GRCh38\_to\_NCBI36.chain.gz/download}{GRCh38\_to\_NCBI36.chain.gz}

\end{itemize}
\begin{itemize}
\item {} 
GRCh37 \textless{}=\textgreater{} GRCh38

\end{itemize}
\begin{itemize}
\item {} 
\sphinxhref{https://sourceforge.net/projects/crossmap/files/Ensembl\_chain\_files/homo\_sapiens\%28human\%29/GRCh37\_to\_GRCh38.chain.gz/download}{GRCh37\_to\_GRCh38.chain.gz}

\item {} 
\sphinxhref{https://sourceforge.net/projects/crossmap/files/Ensembl\_chain\_files/homo\_sapiens\%28human\%29/GRCh38\_to\_GRCh37.chain.gz/download}{GRCh38\_to\_GRCh37.chain.gz}

\end{itemize}
\begin{itemize}
\item {} 
NCBI34 \textless{}=\textgreater{} GRCh37

\end{itemize}
\begin{itemize}
\item {} 
\sphinxhref{https://sourceforge.net/projects/crossmap/files/Ensembl\_chain\_files/homo\_sapiens\%28human\%29/NCBI34\_to\_GRCh37.chain.gz/download}{NCBI34\_to\_GRCh37.chain.gz}

\item {} 
\sphinxhref{https://sourceforge.net/projects/crossmap/files/Ensembl\_chain\_files/homo\_sapiens\%28human\%29/GRCh37\_to\_NCBI34.chain.gz/download}{GRCh37\_to\_NCBI34.chain.gz}

\end{itemize}
\begin{itemize}
\item {} 
NCBI35 \textless{}=\textgreater{} GRCh37

\end{itemize}
\begin{itemize}
\item {} 
\sphinxhref{https://sourceforge.net/projects/crossmap/files/Ensembl\_chain\_files/homo\_sapiens\%28human\%29/NCBI35\_to\_GRCh37.chain.gz/download}{NCBI35\_to\_GRCh37.chain.gz}

\item {} 
\sphinxhref{https://sourceforge.net/projects/crossmap/files/Ensembl\_chain\_files/homo\_sapiens\%28human\%29/GRCh37\_to\_NCBI35.chain.gz/download}{GRCh37\_to\_NCBI35.chain.gz}

\end{itemize}
\begin{itemize}
\item {} 
NCBI36 \textless{}=\textgreater{} GRCh37

\end{itemize}
\begin{itemize}
\item {} 
\sphinxhref{https://sourceforge.net/projects/crossmap/files/Ensembl\_chain\_files/homo\_sapiens\%28human\%29/NCBI36\_to\_GRCh37.chain.gz/download}{NCBI36\_to\_GRCh37.chain.gz}

\item {} 
\sphinxhref{https://sourceforge.net/projects/crossmap/files/Ensembl\_chain\_files/homo\_sapiens\%28human\%29/GRCh37\_to\_NCBI36.chain.gz/download}{GRCh37\_to\_NCBI36.chain.gz}

\end{itemize}

\sphinxstylestrong{Ensembl built chain files (Mouse, Mus musculus)}
\begin{itemize}
\item {} 
\sphinxhref{https://sourceforge.net/projects/crossmap/files/Ensembl\_chain\_files/mus\_musculus\%28mouse\%29/NCBIM37\_to\_GRCm38.chain.gz/download}{NCBIM37\_to\_GRCm38.chain.gz}

\item {} 
\sphinxhref{https://sourceforge.net/projects/crossmap/files/Ensembl\_chain\_files/mus\_musculus\%28mouse\%29/GRCm38\_to\_NCBIM36.chain.gz/download}{GRCm38\_to\_NCBIM36.chain.gz}

\item {} 
\sphinxhref{https://sourceforge.net/projects/crossmap/files/Ensembl\_chain\_files/mus\_musculus\%28mouse\%29/GRCm38\_to\_NCBIM37.chain.gz/download}{GRCm38\_to\_NCBIM37.chain.gz}

\item {} 
\sphinxhref{https://sourceforge.net/projects/crossmap/files/Ensembl\_chain\_files/mus\_musculus\%28mouse\%29/NCBIM36\_to\_GRCm38.chain.gz/download}{NCBIM36\_to\_GRCm38.chain.gz}

\end{itemize}

\sphinxstylestrong{Ensembl Chain file of other species can be downloaded from:} \sphinxurl{ftp://ftp.ensembl.org/pub/assembly\_mapping/}


\section{User Input file}
\label{\detokenize{index:user-input-file}}\begin{enumerate}
\item {} 
\sphinxhref{http://samtools.sourceforge.net/SAMv1.pdf}{BAM} or \sphinxhref{http://samtools.sourceforge.net/SAMv1.pdf/}{SAM} format.

\item {} 
\sphinxhref{http://genome.ucsc.edu/FAQ/FAQformat.html\#format1}{BED} or BED-like format. BED file must has at least 3 columns (`chrom', `start', `end').

\item {} 
\sphinxhref{http://genome.ucsc.edu/goldenPath/help/wiggle.html}{Wiggle} format. \sphinxquotedblleft{}variableStep\sphinxquotedblright{}, \sphinxquotedblleft{}fixedStep\sphinxquotedblright{} and \sphinxquotedblleft{}bedGraph\sphinxquotedblright{} wiggle line are supported.

\item {} 
\sphinxhref{http://genome.ucsc.edu/goldenPath/help/bigWig.html}{BigWig} format.

\item {} 
\sphinxhref{http://genome.ucsc.edu/FAQ/FAQformat.html\#format3}{GFF} or \sphinxhref{http://genome.ucsc.edu/FAQ/FAQformat.html\#format4}{GTF} format.

\item {} 
\sphinxhref{http://vcftools.sourceforge.net/index.html}{VCF} format.

\end{enumerate}

\sphinxstylestrong{NOTE:} When converting \sphinxstylestrong{bedGraph} file, Treat it as \sphinxstylestrong{Wiggle} format rather than \sphinxstylestrong{BED} format.


\section{Output file}
\label{\detokenize{index:output-file}}
Format of Output files depends on the input format (version \textless{}= 0.2.9)

\noindent\begin{tabulary}{\linewidth}{|L|L|}
\hline
\sphinxstylethead{\relax 
Input\_format
\unskip}\relax &\sphinxstylethead{\relax 
Output\_format
\unskip}\relax \\
\hline
BED
&
BED (Genome coordinates will be updated to the target assembly)
\\
\hline
BAM
&
BAM (Genome coordinates, header section, all SAM flags, insert size will be updated accordingly)
\\
\hline
SAM
&
SAM (Genome coordinates, header section, all SAM flags, insert size will be updated accordingly)
\\
\hline
Wiggle
&
bedGraph (if wigToBigWig executable does not exist)
\\
\hline
Wiggle
&
BigWig (if wigToBigWig executable exists)
\\
\hline
BigWig
&
bedGraph (if wigToBigWig executable does not exist)
\\
\hline
BigWig
&
BigWig (if wigToBigWig executable exists)
\\
\hline
GFF
&
GFF (Genome coordinates will be updated to the target assembly)
\\
\hline
GTF
&
GTF (Genome coordinates will be updated to the target assembly)
\\
\hline
VCF
&
VCF (Genome coordinates and reference alleles will be updated to the target assembly)
\\
\hline\end{tabulary}


Format of Output files depends on the input format (version \textgreater{}= 0.3.0)

\noindent\begin{tabulary}{\linewidth}{|L|L|}
\hline
\sphinxstylethead{\relax 
Input\_format
\unskip}\relax &\sphinxstylethead{\relax 
Output\_format
\unskip}\relax \\
\hline
BED
&
BED (Genome coordinates will be updated to the target assembly)
\\
\hline
BAM
&
BAM (Genome coordinates, header section, all SAM flags, insert size will be updated accordingly)
\\
\hline
SAM
&
SAM (Genome coordinates, header section, all SAM flags, insert size will be updated accordingly)
\\
\hline
Wiggle
&
BigWig
\\
\hline
BigWig
&
BigWig
\\
\hline
GFF
&
GFF (Genome coordinates will be updated to the target assembly)
\\
\hline
GTF
&
GTF (Genome coordinates will be updated to the target assembly)
\\
\hline
VCF
&
VCF (Genome coordinates and reference alleles will be updated to the target assembly)
\\
\hline\end{tabulary}



\chapter{Usage}
\label{\detokenize{index:usage}}
Run CrossMap.py without any arguments will print help message:

\begin{sphinxVerbatim}[commandchars=\\\{\}]
\PYGZsh{} run CrossMap without argument
\PYGZdl{} python CrossMap.py
\end{sphinxVerbatim}

Screen output:

\begin{sphinxVerbatim}[commandchars=\\\{\}]
\PYG{n}{Program}\PYG{p}{:} \PYG{n}{CrossMap} \PYG{p}{(}\PYG{n}{v0}\PYG{o}{.}\PYG{l+m+mf}{1.1}\PYG{p}{)}

\PYG{n}{Description}\PYG{p}{:}
  \PYG{n}{CrossMap} \PYG{o+ow}{is} \PYG{n}{a} \PYG{n}{program} \PYG{k}{for} \PYG{n}{convenient} \PYG{n}{conversion} \PYG{n}{of} \PYG{n}{genome} \PYG{n}{coordinates}
  \PYG{o+ow}{and} \PYG{n}{genomeannotation} \PYG{n}{files} \PYG{n}{between} \PYG{n}{assemblies} \PYG{p}{(}\PYG{n}{eg}\PYG{o}{.} \PYG{n}{lift} \PYG{k+kn}{from} \PYG{n+nn}{human}
  \PYG{n}{hg18} \PYG{n}{to} \PYG{n}{hg19} \PYG{o+ow}{or} \PYG{n}{vice} \PYG{n}{versa}\PYG{p}{)}\PYG{o}{.}\PYG{n}{It} \PYG{n}{support} \PYG{n}{file} \PYG{o+ow}{in} \PYG{n}{BAM}\PYG{p}{,} \PYG{n}{SAM}\PYG{p}{,} \PYG{n}{BED}\PYG{p}{,} \PYG{n}{Wiggle}\PYG{p}{,}
  \PYG{n}{BigWig}\PYG{p}{,} \PYG{n}{GFF}\PYG{p}{,} \PYG{n}{GTF}\PYG{p}{,} \PYG{n}{VCF}\PYG{p}{,} \PYG{n}{etc}\PYG{o}{.}

\PYG{n}{Usage}\PYG{p}{:} \PYG{n}{CrossMap}\PYG{o}{.}\PYG{n}{py} \PYG{o}{\PYGZlt{}}\PYG{n}{command}\PYG{o}{\PYGZgt{}} \PYG{p}{[}\PYG{n}{options}\PYG{p}{]}

  \PYG{n}{bam}  \PYG{n}{convert} \PYG{n}{alignment} \PYG{n}{file} \PYG{o+ow}{in} \PYG{n}{BAM} \PYG{o+ow}{or} \PYG{n}{SAM} \PYG{n+nb}{format}\PYG{o}{.}
  \PYG{n}{bed}  \PYG{n}{convert} \PYG{n}{genome} \PYG{n}{cooridnate} \PYG{o+ow}{or} \PYG{n}{annotation} \PYG{n}{file} \PYG{o+ow}{in} \PYG{n}{BED} \PYG{o+ow}{or} \PYG{n}{BED}\PYG{o}{\PYGZhy{}}\PYG{n}{like} \PYG{n+nb}{format}\PYG{o}{.}
  \PYG{n}{bigwig}       \PYG{n}{convert} \PYG{n}{genome} \PYG{n}{coordinate} \PYG{n}{file} \PYG{o+ow}{in} \PYG{n}{BigWig} \PYG{n+nb}{format}\PYG{o}{.}
  \PYG{n}{gff}  \PYG{n}{convert} \PYG{n}{genome} \PYG{n}{cooridnate} \PYG{o+ow}{or} \PYG{n}{annotation} \PYG{n}{file} \PYG{o+ow}{in} \PYG{n}{GFF} \PYG{o+ow}{or} \PYG{n}{GTF} \PYG{n+nb}{format}\PYG{o}{.}
  \PYG{n}{vcf}  \PYG{n}{convert} \PYG{n}{genome} \PYG{n}{coordinate} \PYG{n}{file} \PYG{o+ow}{in} \PYG{n}{VCF} \PYG{n+nb}{format}\PYG{o}{.}
  \PYG{n}{wig}  \PYG{n}{convert} \PYG{n}{genome} \PYG{n}{coordinate} \PYG{n}{file} \PYG{o+ow}{in} \PYG{n}{Wiggle}\PYG{p}{,} \PYG{o+ow}{or} \PYG{n}{bedGraph} \PYG{n+nb}{format}\PYG{o}{.}
\end{sphinxVerbatim}

Run CrossMap.py with command keyword will print help message for that command:

\begin{sphinxVerbatim}[commandchars=\\\{\}]
\PYGZdl{} python CrossMap.py bed
\end{sphinxVerbatim}

Screen output:

\begin{sphinxVerbatim}[commandchars=\\\{\}]
\PYG{n}{Usage}\PYG{p}{:}
  \PYG{n}{CrossMap}\PYG{o}{.}\PYG{n}{py} \PYG{n}{bed} \PYG{n}{input\PYGZus{}chain\PYGZus{}file} \PYG{n}{input\PYGZus{}bed\PYGZus{}file} \PYG{p}{[}\PYG{n}{output\PYGZus{}file}\PYG{p}{]}

\PYG{n}{Description}\PYG{p}{:}
  \PYG{l+s+s2}{\PYGZdq{}}\PYG{l+s+s2}{input\PYGZus{}chain\PYGZus{}file}\PYG{l+s+s2}{\PYGZdq{}} \PYG{o+ow}{and} \PYG{l+s+s2}{\PYGZdq{}}\PYG{l+s+s2}{input\PYGZus{}bed\PYGZus{}file}\PYG{l+s+s2}{\PYGZdq{}} \PYG{n}{can} \PYG{n}{be} \PYG{n}{regular} \PYG{o+ow}{or} \PYG{n}{compressed}
  \PYG{p}{(}\PYG{o}{*}\PYG{o}{.}\PYG{n}{gz}\PYG{p}{,} \PYG{o}{*}\PYG{o}{.}\PYG{n}{Z}\PYG{p}{,} \PYG{o}{*}\PYG{o}{.}\PYG{n}{z}\PYG{p}{,} \PYG{o}{*}\PYG{o}{.}\PYG{n}{bz}\PYG{p}{,} \PYG{o}{*}\PYG{o}{.}\PYG{n}{bz2}\PYG{p}{,} \PYG{o}{*}\PYG{o}{.}\PYG{n}{bzip2}\PYG{p}{)} \PYG{n}{file}\PYG{p}{,} \PYG{n}{local} \PYG{n}{file} \PYG{o+ow}{or} \PYG{n}{URL}
  \PYG{p}{(}\PYG{n}{http}\PYG{p}{:}\PYG{o}{/}\PYG{o}{/}\PYG{p}{,} \PYG{n}{https}\PYG{p}{:}\PYG{o}{/}\PYG{o}{/}\PYG{p}{,} \PYG{n}{ftp}\PYG{p}{:}\PYG{o}{/}\PYG{o}{/}\PYG{p}{)} \PYG{n}{pointing} \PYG{n}{to} \PYG{n}{remote} \PYG{n}{file}\PYG{o}{.} \PYG{n}{BED} \PYG{n}{file} \PYG{n}{must}
  \PYG{n}{have} \PYG{n}{at} \PYG{n}{least} \PYG{l+m+mi}{3} \PYG{n}{columns} \PYG{p}{(}\PYG{n}{chrom}\PYG{p}{,} \PYG{n}{start}\PYG{p}{,} \PYG{n}{end}\PYG{p}{)} \PYG{o+ow}{and} \PYG{n}{no} \PYG{n}{more} \PYG{n}{than} \PYG{l+m+mi}{12}
  \PYG{n}{columns}\PYG{o}{.} \PYG{n}{If}  \PYG{n}{no} \PYG{l+s+s2}{\PYGZdq{}}\PYG{l+s+s2}{output\PYGZus{}file}\PYG{l+s+s2}{\PYGZdq{}} \PYG{n}{was} \PYG{n}{specified}\PYG{p}{,} \PYG{n}{output} \PYG{n}{will} \PYG{n}{be} \PYG{n}{directed}
  \PYG{n}{to} \PYG{n}{screen} \PYG{p}{(}\PYG{n}{console}\PYG{p}{)}\PYG{o}{.} \PYG{n}{BED} \PYG{n+nb}{format}\PYG{p}{:}
  \PYG{n}{http}\PYG{p}{:}\PYG{o}{/}\PYG{o}{/}\PYG{n}{genome}\PYG{o}{.}\PYG{n}{ucsc}\PYG{o}{.}\PYG{n}{edu}\PYG{o}{/}\PYG{n}{FAQ}\PYG{o}{/}\PYG{n}{FAQformat}\PYG{o}{.}\PYG{n}{html}\PYG{c+c1}{\PYGZsh{}format1}

\PYG{n}{Example}\PYG{p}{:}
  \PYG{n}{CrossMapy}\PYG{o}{.}\PYG{n}{py} \PYG{n}{bed} \PYG{n}{hg18ToHg19}\PYG{o}{.}\PYG{n}{over}\PYG{o}{.}\PYG{n}{chain}\PYG{o}{.}\PYG{n}{gz} \PYG{n}{test}\PYG{o}{.}\PYG{n}{hg18}\PYG{o}{.}\PYG{n}{bed} \PYG{n}{test}\PYG{o}{.}\PYG{n}{hg19}\PYG{o}{.}\PYG{n}{bed}
  \PYG{c+c1}{\PYGZsh{} write output to \PYGZdq{}test.hg19.bed\PYGZdq{}}

\PYG{n}{Example}\PYG{p}{:}
  \PYG{n}{CrossMapy}\PYG{o}{.}\PYG{n}{py} \PYG{n}{bed} \PYG{n}{hg18ToHg19}\PYG{o}{.}\PYG{n}{over}\PYG{o}{.}\PYG{n}{chain}\PYG{o}{.}\PYG{n}{gz} \PYG{n}{test}\PYG{o}{.}\PYG{n}{hg18}\PYG{o}{.}\PYG{n}{bed}
  \PYG{c+c1}{\PYGZsh{} write output to screen}
\end{sphinxVerbatim}


\section{Convert BED format files}
\label{\detokenize{index:convert-bed-format-files}}
A \sphinxhref{http://genome.ucsc.edu/FAQ/FAQformat.html\#format1}{BED} (Browser Extensible Data) file
is a tab-delimited text file describing genome regions or gene annotations. It is the standard
file format used by UCSC. It consists of one line per feature, each containing 3-12 columns.
CrossMap converts BED files with less than 12 columns to a different assembly by updating the
chromosome and genome coordinates only; all other columns remain unchanged. Regions from old
assembly mapping to multiple locations to the new assembly will be split.  For 12-columns BED
files, all columns will be updated accordingly except the 4th column (name of bed line), 5th
column (score value) and 9th column (RGB value describing the display color). 12-column BED
files usually define multiple blocks (eg. exon); if any of the exons fails to map to a new
assembly, the whole BED line is skipped.

The input BED file can be plain text file, compressed file with extension of .gz, .Z, .z,
.bz, .bz2 and .bzip2, or even a URL pointing to accessible remote files (\sphinxurl{http://}, \sphinxurl{https://}
and \sphinxurl{ftp://}). Compressed remote files are not supported. The output is a BED format file with
exact the same number of columns as the original one.

Standard \sphinxhref{http://genome.ucsc.edu/FAQ/FAQformat.html\#format1}{BED} format has 12 columns, but CrossMap also supports BED-like formats:
\begin{itemize}
\item {} 
BED3: The first 3 columns (\sphinxquotedblleft{}chrom\sphinxquotedblright{}, \sphinxquotedblleft{}start\sphinxquotedblright{}, \sphinxquotedblleft{}end\sphinxquotedblright{}) of BED format file.

\item {} 
BED6: The first 6 columns (\sphinxquotedblleft{}chrom\sphinxquotedblright{}, \sphinxquotedblleft{}start\sphinxquotedblright{}, \sphinxquotedblleft{}end\sphinxquotedblright{}, \sphinxquotedblleft{}name\sphinxquotedblright{}, \sphinxquotedblleft{}score\sphinxquotedblright{}, \sphinxquotedblleft{}strand\sphinxquotedblright{}) of BED format file.

\item {} 
Other: Format has at least 3 columns (\sphinxquotedblleft{}chrom\sphinxquotedblright{}, \sphinxquotedblleft{}start\sphinxquotedblright{}, \sphinxquotedblleft{}end\sphinxquotedblright{}) and no more than 12 columns. All other columns are arbitrary.

\end{itemize}

NOTE:
\begin{enumerate}
\item {} 
For BED-like formats mentioned above, CrossMap only updates \sphinxquotedblleft{}chrom (1st column)\sphinxquotedblright{}, \sphinxquotedblleft{}start (2nd column) \sphinxquotedblright{}, \sphinxquotedblleft{}end (3rd column) \sphinxquotedblright{} and \sphinxquotedblleft{}strand\sphinxquotedblright{} (if any). All other columns will keep AS-IS.

\item {} 
Lines starting with `\#', `browser', `track' will be skipped.

\item {} 
Lines will less than 3 columns will be skipped.

\item {} 
2nd-column and 3-column must be integer, otherwise skipped.

\item {} 
\sphinxquotedblleft{}+\sphinxquotedblright{} strand is assumed if no strand information was found.

\item {} 
For standard BED format (12 columns). If any of the defined exon blocks cannot be uniquely mapped to target assembly, the whole entry will be skipped.

\item {} 
\sphinxquotedblleft{}input\_chain\_file\sphinxquotedblright{} and \sphinxquotedblleft{}input\_bed\_file\sphinxquotedblright{} can be regular or compressed (.gz, .Z, .z, .bz, .bz2, .bzip2) file, local file or URL (\sphinxurl{http://}, \sphinxurl{https://}, \sphinxurl{ftp://}) pointing to remote file.

\item {} 
If output\_file was not specified, results will be printed to screen (console). In this case, the original bed entries (include items failed to convert) were also printed out.

\item {} 
If input region cannot be consecutively mapped target assembly, it will be split.

\item {} 
*.unmap file contains regions that cannot be unambiguously converted.

\end{enumerate}

Example (run CrossMap with \sphinxstylestrong{no} \sphinxstyleemphasis{output\_file} specified):

\begin{sphinxVerbatim}[commandchars=\\\{\}]
\PYGZdl{} python CrossMap.py bed hg18ToHg19.over.chain.gz test.hg18.bed3
\end{sphinxVerbatim}

Conversion results were printed to screen directly (column1-3 are hg18 based, column5-7 are hg19 based):

\begin{sphinxVerbatim}[commandchars=\\\{\}]
\PYG{n}{chr1}   \PYG{l+m+mi}{142614848}       \PYG{l+m+mi}{142617697}       \PYG{o}{\PYGZhy{}}\PYG{o}{\PYGZgt{}}      \PYG{n}{chr1}    \PYG{l+m+mi}{143903503}       \PYG{l+m+mi}{143906352}
\PYG{n}{chr1}   \PYG{l+m+mi}{142617697}       \PYG{l+m+mi}{142623312}       \PYG{o}{\PYGZhy{}}\PYG{o}{\PYGZgt{}}      \PYG{n}{chr1}    \PYG{l+m+mi}{143906355}       \PYG{l+m+mi}{143911970}
\PYG{n}{chr1}   \PYG{l+m+mi}{142623313}       \PYG{l+m+mi}{142623350}       \PYG{o}{\PYGZhy{}}\PYG{o}{\PYGZgt{}}      \PYG{n}{chr1}    \PYG{l+m+mi}{143911971}       \PYG{l+m+mi}{143912008}
\PYG{n}{chr1}   \PYG{l+m+mi}{142623351}       \PYG{l+m+mi}{142626523}       \PYG{o}{\PYGZhy{}}\PYG{o}{\PYGZgt{}}      \PYG{n}{chr1}    \PYG{l+m+mi}{143912009}       \PYG{l+m+mi}{143915181}
\PYG{n}{chr1}   \PYG{l+m+mi}{142633862}       \PYG{l+m+mi}{142633883}       \PYG{o}{\PYGZhy{}}\PYG{o}{\PYGZgt{}}      \PYG{n}{chr1}    \PYG{l+m+mi}{143922520}       \PYG{l+m+mi}{143922541}
\PYG{n}{chr1}   \PYG{l+m+mi}{142633884}       \PYG{l+m+mi}{142636152}       \PYG{o}{\PYGZhy{}}\PYG{o}{\PYGZgt{}}      \PYG{n}{chr1}    \PYG{l+m+mi}{143922542}       \PYG{l+m+mi}{143924810}
\PYG{n}{chr1}   \PYG{l+m+mi}{142636152}       \PYG{l+m+mi}{142636326}       \PYG{o}{\PYGZhy{}}\PYG{o}{\PYGZgt{}}      \PYG{n}{chr1}    \PYG{l+m+mi}{143924813}       \PYG{l+m+mi}{143924987}
\PYG{n}{chr1}   \PYG{l+m+mi}{142636339}       \PYG{l+m+mi}{142636391}       \PYG{o}{\PYGZhy{}}\PYG{o}{\PYGZgt{}}      \PYG{n}{chr1}    \PYG{l+m+mi}{143925000}       \PYG{l+m+mi}{143925052}
\PYG{n}{chr1}   \PYG{l+m+mi}{142636392}       \PYG{l+m+mi}{142637362}       \PYG{o}{\PYGZhy{}}\PYG{o}{\PYGZgt{}}      \PYG{n}{chr1}    \PYG{l+m+mi}{143925052}       \PYG{l+m+mi}{143926022}
\PYG{n}{chr1}   \PYG{l+m+mi}{142637373}       \PYG{l+m+mi}{142639738}       \PYG{o}{\PYGZhy{}}\PYG{o}{\PYGZgt{}}      \PYG{n}{chr1}    \PYG{l+m+mi}{143926033}       \PYG{l+m+mi}{143928398}
\PYG{n}{chr1}   \PYG{l+m+mi}{142639739}       \PYG{l+m+mi}{142639760}       \PYG{o}{\PYGZhy{}}\PYG{o}{\PYGZgt{}}      \PYG{n}{chr1}    \PYG{l+m+mi}{143928399}       \PYG{l+m+mi}{143928420}
\PYG{n}{chr1}   \PYG{l+m+mi}{142639761}       \PYG{l+m+mi}{142640145}       \PYG{o}{\PYGZhy{}}\PYG{o}{\PYGZgt{}}      \PYG{n}{chr1}    \PYG{l+m+mi}{143928421}       \PYG{l+m+mi}{143928805}
\PYG{n}{chr1}   \PYG{l+m+mi}{142640153}       \PYG{l+m+mi}{142641149}       \PYG{o}{\PYGZhy{}}\PYG{o}{\PYGZgt{}}      \PYG{n}{chr1}    \PYG{l+m+mi}{143928813}       \PYG{l+m+mi}{143929809}
\end{sphinxVerbatim}

Example (run CrossMap with \sphinxstyleemphasis{output\_file} \sphinxstylestrong{(test.hg19.bed3)} specified):

\begin{sphinxVerbatim}[commandchars=\\\{\}]
\PYGZdl{} python CrossMap.py bed hg18ToHg19.over.chain.gz test.hg18.bed3 test.hg19.bed3

\PYGZdl{} cat test.hg19.bed3
chr1   143903503       143906352
chr1   143906355       143911970
chr1   143911971       143912008
chr1   143912009       143915181
chr1   143922520       143922541
chr1   143922542       143924810
chr1   143924813       143924987
chr1   143925000       143925052
chr1   143925052       143926022
chr1   143926033       143928398
chr1   143928399       143928420
chr1   143928421       143928805
chr1   143928813       143929809
\end{sphinxVerbatim}

Example (one input region was split because it cannot be consecutively mapped target assembly):

\begin{sphinxVerbatim}[commandchars=\\\{\}]
\PYGZdl{} python CrossMap.py bed hg18ToHg19.over.chain.gz test.hg18.bed3

chr10  81346644        81349952        +       \PYGZhy{}\PYGZgt{}      chr10   81356692        81360000        +
chr10  81349952        81364937        +       \PYGZhy{}\PYGZgt{}      chr10   81360000        81374985        +
chr10  81364952        81365854        +       \PYGZhy{}\PYGZgt{}      chr10   81375000        81375902        +
chr10  81365875        81369946        +       \PYGZhy{}\PYGZgt{}      chr10   81375929        81380000        +
chr10  81369946        81370453        +       \PYGZhy{}\PYGZgt{}      chr10   81380000        81380507        +
chr10  81370483        81371363        +       \PYGZhy{}\PYGZgt{}      chr10   81380539        81381419        +
chr10  81371363        81371365        +       \PYGZhy{}\PYGZgt{}      chr10   62961832        62961834        +
chr10  81371412        81371432        +       (split.1:chr10:81371412:81371422:+)     chr10   62961775        62961785        +
chr10  81371412        81371432        +       (split.2:chr10:81371422:81371432:+)     chrX    63278348        63278358        +
\end{sphinxVerbatim}

Example (Use \sphinxstylestrong{bed} command to convert a bedGraph file, output another bedGraph file. If Use \sphinxstylestrong{wig} command to convert a bedGraph file, output a \sphinxstylestrong{bigWig} file. ):

\begin{sphinxVerbatim}[commandchars=\\\{\}]
\PYGZdl{} python3 ../bin/CrossMap.py bed ../data/UCSC\PYGZus{}chain/hg19ToHg38.over.chain.gz 4\PYGZus{}hg19.bgr

chrX   5873316 5873391 2.0     \PYGZhy{}\PYGZgt{}      chrX    5955275 5955350 2.0
chrX   5873673 5873710 0.8     \PYGZhy{}\PYGZgt{}      chrX    5955632 5955669 0.8
chrX   5873710 5873785 1.4     \PYGZhy{}\PYGZgt{}      chrX    5955669 5955744 1.4
chrX   5873896 5873929 0.9     \PYGZhy{}\PYGZgt{}      chrX    5955855 5955888 0.9
chrX   5873929 5874004 1.5     \PYGZhy{}\PYGZgt{}      chrX    5955888 5955963 1.5
chrX   5874230 5874471 0.3     \PYGZhy{}\PYGZgt{}      chrX    5956189 5956430 0.3
chrX   5874471 5874518 0.9     \PYGZhy{}\PYGZgt{}      chrX    5956430 5956477 0.9

\PYGZdl{} python3 ../bin/CrossMap.py wig ../data/UCSC\PYGZus{}chain/hg19ToHg38.over.chain.gz 4\PYGZus{}hg19.bgr output\PYGZus{}hg38
@ 2018\PYGZhy{}11\PYGZhy{}06 00:09:11: Read chain\PYGZus{}file:  ../data/UCSC\PYGZus{}chain/hg19ToHg38.over.chain.gz
@ 2018\PYGZhy{}11\PYGZhy{}06 00:09:12: Liftover wiggle file: 4\PYGZus{}hg19.bgr ==\PYGZgt{} output\PYGZus{}hg38.bgr
@ 2018\PYGZhy{}11\PYGZhy{}06 00:09:12: Merging overlapped entries in bedGraph file ...
@ 2018\PYGZhy{}11\PYGZhy{}06 00:09:12: Sorting bedGraph file:output\PYGZus{}hg38.bgr
@ 2018\PYGZhy{}11\PYGZhy{}06 00:09:12: Writing header to \PYGZdq{}output\PYGZus{}hg38.bw\PYGZdq{} ...
@ 2018\PYGZhy{}11\PYGZhy{}06 00:09:12: Writing entries to \PYGZdq{}output\PYGZus{}hg38.bw\PYGZdq{} ...
\end{sphinxVerbatim}


\section{Convert BAM/SAM format files}
\label{\detokenize{index:bam-conversion}}\label{\detokenize{index:convert-bam-sam-format-files}}
\sphinxhref{http://samtools.sourceforge.net/samtools.shtml\#5}{SAM} (Sequence Alignment Map) format
is a generic format for storing sequencing alignments, and BAM is binary and compressed
version of SAM (\sphinxhref{http://bioinformatics.oxfordjournals.org/content/25/16/2078.full}{Li et al., 2009}).
Most high-throughput sequencing  (HTS) alignments were in SAM/BAM format and many HTS analysis
tools work with SAM/BAM format. CrossMap updates chromosomes, genome coordinates, header
sections, and all SAM flags accordingly.  The program version (of CrossMap) is inserted into
the header section, along with  the names of the original BAM file and the chain file.  For
pair-end sequencing, insert size is also recalculated. The input BAM file should be sorted
and indexed properly using samTools (\sphinxhref{http://bioinformatics.oxfordjournals.org/content/25/16/2078.full}{Li et al., 2009}).
Output format is determined from the input format and BAM output will be sorted and indexed automatically.

Typing command without any arguments will print help message:

\begin{sphinxVerbatim}[commandchars=\\\{\}]
\PYGZdl{} python CrossMap.py bam
\end{sphinxVerbatim}

Screen output:

\begin{sphinxVerbatim}[commandchars=\\\{\}]
\PYG{n}{Usage}\PYG{p}{:} \PYG{n}{CrossMap}\PYG{o}{.}\PYG{n}{py} \PYG{n}{bam} \PYG{n}{input\PYGZus{}chain\PYGZus{}file} \PYG{n}{input\PYGZus{}bam\PYGZus{}file} \PYG{n}{output\PYGZus{}file} \PYG{p}{[}\PYG{n}{options}\PYG{p}{]}
\PYG{n}{Note}\PYG{p}{:} \PYG{n}{If} \PYG{n}{output\PYGZus{}file} \PYG{o}{==} \PYG{n}{STDOUT} \PYG{o+ow}{or} \PYG{o}{\PYGZhy{}}\PYG{p}{,} \PYG{n}{CrossMap} \PYG{n}{will} \PYG{n}{write} \PYG{n}{BAM} \PYG{n}{file} \PYG{n}{to} \PYG{n}{the} \PYG{n}{screen}

\PYG{n}{Options}\PYG{p}{:}
  \PYG{o}{\PYGZhy{}}\PYG{n}{m} \PYG{n}{INSERT\PYGZus{}SIZE}\PYG{p}{,} \PYG{o}{\PYGZhy{}}\PYG{o}{\PYGZhy{}}\PYG{n}{mean}\PYG{o}{=}\PYG{n}{INSERT\PYGZus{}SIZE}
                       \PYG{n}{Average} \PYG{n}{insert} \PYG{n}{size} \PYG{n}{of} \PYG{n}{pair}\PYG{o}{\PYGZhy{}}\PYG{n}{end} \PYG{n}{sequencing} \PYG{p}{(}\PYG{n}{bp}\PYG{p}{)}\PYG{o}{.}
                       \PYG{p}{[}\PYG{n}{default}\PYG{o}{=}\PYG{l+m+mf}{200.0}\PYG{p}{]}
  \PYG{o}{\PYGZhy{}}\PYG{n}{s} \PYG{n}{INSERT\PYGZus{}SIZE\PYGZus{}STDEV}\PYG{p}{,} \PYG{o}{\PYGZhy{}}\PYG{o}{\PYGZhy{}}\PYG{n}{stdev}\PYG{o}{=}\PYG{n}{INSERT\PYGZus{}SIZE\PYGZus{}STDEV}
                       \PYG{n}{Stanadard} \PYG{n}{deviation} \PYG{n}{of} \PYG{n}{insert} \PYG{n}{size}\PYG{o}{.} \PYG{p}{[}\PYG{n}{default}\PYG{o}{=}\PYG{l+m+mf}{30.0}\PYG{p}{]}
  \PYG{o}{\PYGZhy{}}\PYG{n}{t} \PYG{n}{INSERT\PYGZus{}SIZE\PYGZus{}FOLD}\PYG{p}{,} \PYG{o}{\PYGZhy{}}\PYG{o}{\PYGZhy{}}\PYG{n}{times}\PYG{o}{=}\PYG{n}{INSERT\PYGZus{}SIZE\PYGZus{}FOLD}
                       \PYG{n}{A} \PYG{n}{mapped} \PYG{n}{pair} \PYG{o+ow}{is} \PYG{n}{considered} \PYG{k}{as} \PYG{l+s+s2}{\PYGZdq{}}\PYG{l+s+s2}{proper pair}\PYG{l+s+s2}{\PYGZdq{}} \PYG{k}{if} \PYG{n}{both}
                       \PYG{n}{ends} \PYG{n}{mapped} \PYG{n}{to} \PYG{n}{different} \PYG{n}{strand} \PYG{o+ow}{and} \PYG{n}{the} \PYG{n}{distance}
                       \PYG{n}{between} \PYG{n}{them} \PYG{o+ow}{is} \PYG{n}{less} \PYG{n}{then} \PYG{l+s+s1}{\PYGZsq{}}\PYG{l+s+s1}{\PYGZhy{}t}\PYG{l+s+s1}{\PYGZsq{}} \PYG{o}{*} \PYG{n}{stdev} \PYG{k+kn}{from} \PYG{n+nn}{the} \PYG{n}{mean}\PYG{o}{.}
                       \PYG{p}{[}\PYG{n}{default}\PYG{o}{=}\PYG{l+m+mf}{3.0}\PYG{p}{]}
  \PYG{o}{\PYGZhy{}}\PYG{n}{a}\PYG{p}{,} \PYG{o}{\PYGZhy{}}\PYG{o}{\PYGZhy{}}\PYG{n}{append}\PYG{o}{\PYGZhy{}}\PYG{n}{tags}     \PYG{n}{Add} \PYG{n}{tag} \PYG{n}{to} \PYG{n}{each} \PYG{n}{alignment}\PYG{o}{.}
\end{sphinxVerbatim}

Example (Convert BAM from hg19 to hg18):

\begin{sphinxVerbatim}[commandchars=\\\{\}]
\PYGZsh{} add optional tags using \PYGZsq{}\PYGZhy{}a\PYGZsq{} (recommend always use \PYGZsq{}\PYGZhy{}a\PYGZsq{} option)

\PYGZdl{} CrossMap.py bam \PYGZhy{}a ../data/hg19ToHg18.over.chain.gz test.hg19.bam test.hg18
Insert size = 200.000000
Insert size stdev = 30.000000
Number of stdev from the mean = 3.000000
Add tags to each alignment = True
@ 2016\PYGZhy{}10\PYGZhy{}07 15:29:06: Read chain\PYGZus{}file:  ../data/hg19ToHg18.over.chain.gz
@ 2016\PYGZhy{}10\PYGZhy{}07 15:29:07: Liftover BAM file: test.hg19.bam ==\PYGZgt{} test.hg18.bam
@ 2016\PYGZhy{}10\PYGZhy{}07 15:29:14: Done!
@ 2016\PYGZhy{}10\PYGZhy{}07 15:29:14: Sort \PYGZdq{}test.hg18.bam\PYGZdq{} ...
@ 2016\PYGZhy{}10\PYGZhy{}07 15:29:15: Index \PYGZdq{}test.hg18.sorted.bam\PYGZdq{} ...
Total alignments:99914
       QC failed: 0
       R1 unique, R2 unique (UU): 96094
       R1 unique, R2 unmapp (UN): 3579
       R1 unique, R2 multiple (UM): 0
       R1 multiple, R2 multiple (MM): 0
       R1 multiple, R2 unique (MU): 233
       R1 multiple, R2 unmapped (MN): 8
       R1 unmap, R2 unmap (NN): 0
       R1 unmap, R2 unique (NU): 0
       R1 unmap, R2 multiple (NM): 0
\end{sphinxVerbatim}

\# BAM/SAM header sections was updated:

\begin{sphinxVerbatim}[commandchars=\\\{\}]
\PYGZdl{} samtools view \PYGZhy{}H  test.hg19.bam
@SQ    SN:chr1 LN:249250621
@SQ    SN:chr2 LN:243199373
@SQ    SN:chr3 LN:198022430
@SQ    SN:chr4 LN:191154276
@SQ    SN:chr5 LN:180915260
@SQ    SN:chr6 LN:171115067
@SQ    SN:chr7 LN:159138663
@SQ    SN:chr8 LN:146364022
@SQ    SN:chr9 LN:141213431
@SQ    SN:chr10        LN:135534747
@SQ    SN:chr11        LN:135006516
@SQ    SN:chr12        LN:133851895
@SQ    SN:chr13        LN:115169878
@SQ    SN:chr14        LN:107349540
@SQ    SN:chr15        LN:102531392
@SQ    SN:chr16        LN:90354753
@SQ    SN:chr17        LN:81195210
@SQ    SN:chr18        LN:78077248
@SQ    SN:chr19        LN:59128983
@SQ    SN:chr20        LN:63025520
@SQ    SN:chr21        LN:48129895
@SQ    SN:chr22        LN:51304566
@SQ    SN:chrX LN:155270560
@SQ    SN:chrY LN:59373566
@SQ    SN:chrM LN:16571
@RG    ID:Sample\PYGZus{}618545BE      SM:Sample\PYGZus{}618545BE      LB:Sample\PYGZus{}618545BE      PL:Illumina
@PG    ID:bwa  PN:bwa  VN:0.6.2\PYGZhy{}r126

\PYGZdl{} samtools view \PYGZhy{}H  test.hg18.bam
@HD    VN:1.0  SO:coordinate
@SQ    SN:chr1 LN:247249719
@SQ    SN:chr10        LN:135374737
@SQ    SN:chr11        LN:134452384
@SQ    SN:chr11\PYGZus{}random LN:215294
@SQ    SN:chr12        LN:132349534
@SQ    SN:chr13        LN:114142980
@SQ    SN:chr13\PYGZus{}random LN:186858
@SQ    SN:chr14        LN:106368585
@SQ    SN:chr15        LN:100338915
@SQ    SN:chr15\PYGZus{}random LN:784346
@SQ    SN:chr16        LN:88827254
@SQ    SN:chr17        LN:78774742
@SQ    SN:chr17\PYGZus{}random LN:2617613
@SQ    SN:chr18        LN:76117153
@SQ    SN:chr18\PYGZus{}random LN:4262
@SQ    SN:chr19        LN:63811651
@SQ    SN:chr19\PYGZus{}random LN:301858
@SQ    SN:chr1\PYGZus{}random  LN:1663265
@SQ    SN:chr2 LN:242951149
@SQ    SN:chr20        LN:62435964
@SQ    SN:chr21        LN:46944323
@SQ    SN:chr21\PYGZus{}random LN:1679693
@SQ    SN:chr22        LN:49691432
@SQ    SN:chr22\PYGZus{}random LN:257318
@SQ    SN:chr3 LN:199501827
@SQ    SN:chr3\PYGZus{}random  LN:749256
@SQ    SN:chr4 LN:191273063
@SQ    SN:chr4\PYGZus{}random  LN:842648
@SQ    SN:chr5 LN:180857866
@SQ    SN:chr6 LN:170899992
@SQ    SN:chr6\PYGZus{}random  LN:1875562
@SQ    SN:chr7 LN:158821424
@SQ    SN:chr7\PYGZus{}random  LN:549659
@SQ    SN:chr8 LN:146274826
@SQ    SN:chr8\PYGZus{}random  LN:943810
@SQ    SN:chr9 LN:140273252
@SQ    SN:chr9\PYGZus{}random  LN:1146434
@SQ    SN:chrM LN:16571
@SQ    SN:chrX LN:154913754
@SQ    SN:chrX\PYGZus{}random  LN:1719168
@SQ    SN:chrY LN:57772954
@RG    ID:Sample\PYGZus{}618545BE      SM:Sample\PYGZus{}618545BE      LB:Sample\PYGZus{}618545BE      PL:Illumina
@PG    PN:bwa  ID:bwa  VN:0.6.2\PYGZhy{}r126
@PG    ID:CrossMap     VN:0.1.3
@CO    Liftover from original BAM/SAM file: test.hg19.bam
@CO    Liftover is based on the chain file: ../test/hg19ToHg18.over.chain.gz
\end{sphinxVerbatim}

\sphinxstylestrong{Optional tags:}
\begin{description}
\item[{Q}] \leavevmode
QC. QC failed.

\item[{N}] \leavevmode
Unmapped. Originally unmapped or originally mapped but failed to liftover to new assembly.

\item[{M}] \leavevmode
Multiple mapped. Alignment can be liftover to multiple places.

\item[{U}] \leavevmode
Unique mapped. Alignment can be liftover to only 1 place.

\end{description}

\sphinxstylestrong{Tags for pair-end sequencing include:}
\begin{itemize}
\item {} 
QF = QC failed

\item {} 
NN = both read1 and read2 unmapped

\item {} 
NU = read1 unmapped, read2 unique mapped

\item {} 
NM = read1 unmapped, multiple mapped

\item {} 
UN = read1 uniquely mapped, read2 unmap

\item {} 
UU = both read1 and read2 uniquely mapped

\item {} 
UM = read1 uniquely mapped, read2 multiple mapped

\item {} 
MN = read1 multiple mapped, read2 unmapped

\item {} 
MU = read1 multiple mapped, read2 unique mapped

\item {} 
MM = both read1 and read2 multiple mapped

\end{itemize}

\sphinxstylestrong{Tags for single-end sequencing include:}
\begin{itemize}
\item {} 
QF = QC failed

\item {} 
SN = unmaped

\item {} 
SM = multiple mapped

\item {} 
SU = uniquely mapped

\end{itemize}

NOTE:
\begin{enumerate}
\item {} 
All alignments (mapped, partial mapped, unmapped, QC failed) will write to one file. Users can filter them by tags (this is why `-a' is always recommended).

\item {} 
Header section will be updated to target assembly.

\item {} 
Genome coordinates and all SAM flags in alignment section will be updated to target assembly.

\item {} 
Optional fields in alignment section will not be updated in current version.

\end{enumerate}


\section{Convert Wiggle/BigWig format files}
\label{\detokenize{index:convert-wiggle-bigwig-format-files}}
\sphinxhref{http://genome.ucsc.edu/goldenPath/help/wiggle.html}{Wiggle} (WIG) format is useful for
displaying continuous data such as GC content and reads intensity of high-throughput sequencing data.
BigWig is a self-indexed binary-format Wiggle file, and has the advantage of supporting random access.
This means only regions that need to be displayed are retrieved by genome browser, and it dramatically
reduces the time needed for data transferring (\sphinxhref{http://bioinformatics.oxfordjournals.org/content/26/17/2204.long}{Kent et al., 2010}).
Input wiggle data can be in variableStep (for data with irregular intervals) or fixedStep
(for data with regular intervals). Regardless of the input, the output will always in bedGraph
format. bedGraph format is similar to wiggle format and can be converted into BigWig format
using UCSC \sphinxhref{http://hgdownload.cse.ucsc.edu/admin/exe/}{wigToBigWig} tool. We export files
in bedGraph because it is usually much smaller than file in wiggle format, and more importantly,
CrossMap internally transforms wiggle into bedGraph to increase running speed.

If an input file is in BigWig format, the output is BigWig format if UCSCâ€™s
`\sphinxhref{http://hgdownload.cse.ucsc.edu/admin/exe/}{wigToBigWig}` executable can be found;
otherwise, the output file will be in bedGraph format.

Typing command without any arguments will print help message:

\begin{sphinxVerbatim}[commandchars=\\\{\}]
\PYGZdl{} python2.7 CrossMap.py  wig
\end{sphinxVerbatim}

Screen output:

\begin{sphinxVerbatim}[commandchars=\\\{\}]
\PYG{n}{Usage}\PYG{p}{:}
  \PYG{n}{CrossMap}\PYG{o}{.}\PYG{n}{py} \PYG{n}{wig} \PYG{n}{input\PYGZus{}chain\PYGZus{}file} \PYG{n}{input\PYGZus{}wig\PYGZus{}file} \PYG{n}{output\PYGZus{}prefix}

\PYG{n}{Description}\PYG{p}{:}
  \PYG{l+s+s2}{\PYGZdq{}}\PYG{l+s+s2}{input\PYGZus{}chain\PYGZus{}file}\PYG{l+s+s2}{\PYGZdq{}} \PYG{n}{can} \PYG{n}{be} \PYG{n}{regular} \PYG{o+ow}{or} \PYG{n}{compressed} \PYG{p}{(}\PYG{o}{*}\PYG{o}{.}\PYG{n}{gz}\PYG{p}{,} \PYG{o}{*}\PYG{o}{.}\PYG{n}{Z}\PYG{p}{,} \PYG{o}{*}\PYG{o}{.}\PYG{n}{z}\PYG{p}{,} \PYG{o}{*}\PYG{o}{.}\PYG{n}{bz}\PYG{p}{,} \PYG{o}{*}\PYG{o}{.}\PYG{n}{bz2}\PYG{p}{,}
  \PYG{o}{*}\PYG{o}{.}\PYG{n}{bzip2}\PYG{p}{)} \PYG{n}{file}\PYG{p}{,} \PYG{n}{local} \PYG{n}{file} \PYG{o+ow}{or} \PYG{n}{URL} \PYG{p}{(}\PYG{n}{http}\PYG{p}{:}\PYG{o}{/}\PYG{o}{/}\PYG{p}{,} \PYG{n}{https}\PYG{p}{:}\PYG{o}{/}\PYG{o}{/}\PYG{p}{,} \PYG{n}{ftp}\PYG{p}{:}\PYG{o}{/}\PYG{o}{/}\PYG{p}{)} \PYG{n}{pointing} \PYG{n}{to} \PYG{n}{remote}
  \PYG{n}{file}\PYG{o}{.}  \PYG{n}{Both} \PYG{l+s+s2}{\PYGZdq{}}\PYG{l+s+s2}{variableStep}\PYG{l+s+s2}{\PYGZdq{}} \PYG{o+ow}{and} \PYG{l+s+s2}{\PYGZdq{}}\PYG{l+s+s2}{fixedStep}\PYG{l+s+s2}{\PYGZdq{}} \PYG{n}{wiggle} \PYG{n}{lines} \PYG{n}{are} \PYG{n}{supported}\PYG{o}{.} \PYG{n}{Wiggle}
  \PYG{n+nb}{format}\PYG{p}{:} \PYG{n}{http}\PYG{p}{:}\PYG{o}{/}\PYG{o}{/}\PYG{n}{genome}\PYG{o}{.}\PYG{n}{ucsc}\PYG{o}{.}\PYG{n}{edu}\PYG{o}{/}\PYG{n}{goldenPath}\PYG{o}{/}\PYG{n}{help}\PYG{o}{/}\PYG{n}{wiggle}\PYG{o}{.}\PYG{n}{html}

\PYG{n}{Example}\PYG{p}{:}
  \PYG{n}{CrossMapy}\PYG{o}{.}\PYG{n}{py} \PYG{n}{wig} \PYG{n}{hg18ToHg19}\PYG{o}{.}\PYG{n}{over}\PYG{o}{.}\PYG{n}{chain}\PYG{o}{.}\PYG{n}{gz} \PYG{n}{test}\PYG{o}{.}\PYG{n}{hg18}\PYG{o}{.}\PYG{n}{wig} \PYG{n}{test}\PYG{o}{.}\PYG{n}{hg19}
\end{sphinxVerbatim}

NOTE:
\begin{enumerate}
\item {} 
To improve performance, this script calls \sphinxhref{http://www.gnu.org/software/coreutils/manual/html\_node/sort-invocation.html}{GNU \sphinxquotedblleft{}sort\sphinxquotedblright{}} command internally.
If \sphinxquotedblleft{}sort\sphinxquotedblright{} command does not exist, CrossMap will exit.

\end{enumerate}

Typing command without any arguments will print help message:

\begin{sphinxVerbatim}[commandchars=\\\{\}]
\PYGZdl{} python2.7 CrossMap.py  bigwig
\end{sphinxVerbatim}

Screen output:

\begin{sphinxVerbatim}[commandchars=\\\{\}]
\PYG{n}{Usage}\PYG{p}{:}
  \PYG{n}{CrossMap}\PYG{o}{.}\PYG{n}{py} \PYG{n}{bigwig} \PYG{n}{input\PYGZus{}chain\PYGZus{}file} \PYG{n}{input\PYGZus{}\PYGZus{}bigwig\PYGZus{}file} \PYG{n}{output\PYGZus{}prefix}

\PYG{n}{Description}\PYG{p}{:}
  \PYG{l+s+s2}{\PYGZdq{}}\PYG{l+s+s2}{input\PYGZus{}chain\PYGZus{}file}\PYG{l+s+s2}{\PYGZdq{}} \PYG{n}{can} \PYG{n}{be} \PYG{n}{regular} \PYG{o+ow}{or} \PYG{n}{compressed} \PYG{p}{(}\PYG{o}{*}\PYG{o}{.}\PYG{n}{gz}\PYG{p}{,} \PYG{o}{*}\PYG{o}{.}\PYG{n}{Z}\PYG{p}{,} \PYG{o}{*}\PYG{o}{.}\PYG{n}{z}\PYG{p}{,} \PYG{o}{*}\PYG{o}{.}\PYG{n}{bz}\PYG{p}{,} \PYG{o}{*}\PYG{o}{.}\PYG{n}{bz2}\PYG{p}{,}
  \PYG{o}{*}\PYG{o}{.}\PYG{n}{bzip2}\PYG{p}{)} \PYG{n}{file}\PYG{p}{,} \PYG{n}{local} \PYG{n}{file} \PYG{o+ow}{or} \PYG{n}{URL} \PYG{p}{(}\PYG{n}{http}\PYG{p}{:}\PYG{o}{/}\PYG{o}{/}\PYG{p}{,} \PYG{n}{https}\PYG{p}{:}\PYG{o}{/}\PYG{o}{/}\PYG{p}{,} \PYG{n}{ftp}\PYG{p}{:}\PYG{o}{/}\PYG{o}{/}\PYG{p}{)} \PYG{n}{pointing} \PYG{n}{to} \PYG{n}{remote}
  \PYG{n}{file}\PYG{o}{.} \PYG{n}{Bigwig} \PYG{n+nb}{format}\PYG{p}{:} \PYG{n}{http}\PYG{p}{:}\PYG{o}{/}\PYG{o}{/}\PYG{n}{genome}\PYG{o}{.}\PYG{n}{ucsc}\PYG{o}{.}\PYG{n}{edu}\PYG{o}{/}\PYG{n}{goldenPath}\PYG{o}{/}\PYG{n}{help}\PYG{o}{/}\PYG{n}{bigWig}\PYG{o}{.}\PYG{n}{html}

\PYG{n}{Example}\PYG{p}{:}
  \PYG{n}{CrossMapy}\PYG{o}{.}\PYG{n}{py} \PYG{n}{bigwig} \PYG{n}{hg18ToHg19}\PYG{o}{.}\PYG{n}{over}\PYG{o}{.}\PYG{n}{chain}\PYG{o}{.}\PYG{n}{gz} \PYG{n}{test}\PYG{o}{.}\PYG{n}{hg18}\PYG{o}{.}\PYG{n}{bw} \PYG{n}{test}\PYG{o}{.}\PYG{n}{hg19}
\end{sphinxVerbatim}

Example (Convert BigWig file from hg18 to hg19):

\begin{sphinxVerbatim}[commandchars=\\\{\}]
\PYGZdl{} python CrossMap.py bigwig  hg19ToHg18.over.chain.gz  test.hg19.bw test.hg18
@ 2013\PYGZhy{}11\PYGZhy{}17 22:12:42: Read chain\PYGZus{}file:  ../data/hg19ToHg18.over.chain.gz
@ 2013\PYGZhy{}11\PYGZhy{}17 22:12:44: Liftover bigwig file: test.hg19.bw ==\PYGZgt{} test.hg18.bgr
@ 2013\PYGZhy{}11\PYGZhy{}17 22:15:38: Merging overlapped entries in bedGraph file ...
@ 2013\PYGZhy{}11\PYGZhy{}17 22:15:38: Sorting bedGraph file:test.hg18.bgr
@ 2013\PYGZhy{}11\PYGZhy{}17 22:15:39: Convert wiggle to bigwig ...
\end{sphinxVerbatim}

NOTE:
\begin{enumerate}
\item {} 
To improve performance, this script calls \sphinxhref{http://www.gnu.org/software/coreutils/manual/html\_node/sort-invocation.html}{GNU \sphinxquotedblleft{}sort\sphinxquotedblright{}} command
internally. If \sphinxquotedblleft{}sort\sphinxquotedblright{} command does not exist, CrossMap will exit.

\item {} 
Output files: output\_prefix.bw, output\_prefix.bgr, output\_prefix.sorted.bgr

\end{enumerate}


\section{Convert GFF/GTF format files}
\label{\detokenize{index:convert-gff-gtf-format-files}}
\sphinxhref{http://genome.ucsc.edu/FAQ/FAQformat.html\#format3}{GFF} (General Feature Format) is another
plain text file used to describe gene structure. \sphinxhref{http://genome.ucsc.edu/FAQ/FAQformat.html\#format4}{GTF}
(Gene Transfer Format) is a refined version of GTF. The first eight fields are the same as
GFF. Plain text, compressed plain text, and URLs pointing to remote files are all supported.
Only chromosome and genome coordinates are updated. The format of output is determined from
the input.

Typing command without any arguments will print help message:

\begin{sphinxVerbatim}[commandchars=\\\{\}]
\PYGZdl{} python2.7 CrossMap.py  gff
\end{sphinxVerbatim}

Screen output:

\begin{sphinxVerbatim}[commandchars=\\\{\}]
\PYG{n}{Usage}\PYG{p}{:}
  \PYG{n}{CrossMap}\PYG{o}{.}\PYG{n}{py} \PYG{n}{gff} \PYG{n}{input\PYGZus{}chain\PYGZus{}file} \PYG{n}{input\PYGZus{}gff\PYGZus{}file} \PYG{n}{output\PYGZus{}file}

\PYG{n}{Description}\PYG{p}{:}
  \PYG{l+s+s2}{\PYGZdq{}}\PYG{l+s+s2}{input\PYGZus{}chain\PYGZus{}file}\PYG{l+s+s2}{\PYGZdq{}} \PYG{n}{can} \PYG{n}{be} \PYG{n}{regular} \PYG{o+ow}{or} \PYG{n}{compressed} \PYG{p}{(}\PYG{o}{*}\PYG{o}{.}\PYG{n}{gz}\PYG{p}{,} \PYG{o}{*}\PYG{o}{.}\PYG{n}{Z}\PYG{p}{,} \PYG{o}{*}\PYG{o}{.}\PYG{n}{z}\PYG{p}{,} \PYG{o}{*}\PYG{o}{.}\PYG{n}{bz}\PYG{p}{,} \PYG{o}{*}\PYG{o}{.}\PYG{n}{bz2}\PYG{p}{,}
  \PYG{o}{*}\PYG{o}{.}\PYG{n}{bzip2}\PYG{p}{)} \PYG{n}{file}\PYG{p}{,} \PYG{n}{local} \PYG{n}{file} \PYG{o+ow}{or} \PYG{n}{URL} \PYG{p}{(}\PYG{n}{http}\PYG{p}{:}\PYG{o}{/}\PYG{o}{/}\PYG{p}{,} \PYG{n}{https}\PYG{p}{:}\PYG{o}{/}\PYG{o}{/}\PYG{p}{,} \PYG{n}{ftp}\PYG{p}{:}\PYG{o}{/}\PYG{o}{/}\PYG{p}{)} \PYG{n}{pointing} \PYG{n}{to} \PYG{n}{remote}
  \PYG{n}{file}\PYG{o}{.} \PYG{n+nb}{input} \PYG{n}{file} \PYG{n}{must} \PYG{n}{be} \PYG{o+ow}{in} \PYG{n}{GFF} \PYG{o+ow}{or} \PYG{n}{GTF} \PYG{n+nb}{format}\PYG{o}{.} \PYG{n}{GFF} \PYG{n+nb}{format}\PYG{p}{:}
  \PYG{n}{http}\PYG{p}{:}\PYG{o}{/}\PYG{o}{/}\PYG{n}{genome}\PYG{o}{.}\PYG{n}{ucsc}\PYG{o}{.}\PYG{n}{edu}\PYG{o}{/}\PYG{n}{FAQ}\PYG{o}{/}\PYG{n}{FAQformat}\PYG{o}{.}\PYG{n}{html}\PYG{c+c1}{\PYGZsh{}format3 GTF format:}
  \PYG{n}{http}\PYG{p}{:}\PYG{o}{/}\PYG{o}{/}\PYG{n}{genome}\PYG{o}{.}\PYG{n}{ucsc}\PYG{o}{.}\PYG{n}{edu}\PYG{o}{/}\PYG{n}{FAQ}\PYG{o}{/}\PYG{n}{FAQformat}\PYG{o}{.}\PYG{n}{html}\PYG{c+c1}{\PYGZsh{}format4}

\PYG{n}{Example}\PYG{p}{:}
  \PYG{n}{CrossMap}\PYG{o}{.}\PYG{n}{py} \PYG{n}{gff}  \PYG{n}{hg19ToHg18}\PYG{o}{.}\PYG{n}{over}\PYG{o}{.}\PYG{n}{chain}\PYG{o}{.}\PYG{n}{gz} \PYG{n}{test}\PYG{o}{.}\PYG{n}{hg19}\PYG{o}{.}\PYG{n}{gtf} \PYG{n}{test}\PYG{o}{.}\PYG{n}{hg18}\PYG{o}{.}\PYG{n}{gtf} \PYG{c+c1}{\PYGZsh{}write output to test.hg18.gtf}

\PYG{n}{Example}\PYG{p}{:}
   \PYG{n}{CrossMap}\PYG{o}{.}\PYG{n}{py} \PYG{n}{gff}  \PYG{n}{hg19ToHg18}\PYG{o}{.}\PYG{n}{over}\PYG{o}{.}\PYG{n}{chain}\PYG{o}{.}\PYG{n}{gz} \PYG{n}{test}\PYG{o}{.}\PYG{n}{hg19}\PYG{o}{.}\PYG{n}{gtf}  \PYG{c+c1}{\PYGZsh{} write output to screen}
\end{sphinxVerbatim}

Example (Convert GTF file from hg19 to hg18):

\begin{sphinxVerbatim}[commandchars=\\\{\}]
\PYGZdl{} python CrossMap.py gff  hg19ToHg18.over.chain.gz test.hg19.gtf test.hg18.gtf
@ 2013\PYGZhy{}11\PYGZhy{}17 20:44:47: Read chain\PYGZus{}file:  ../data/hg19ToHg18.over.chain.gz

\PYGZdl{} head test.hg19.gtf
chr1   hg19\PYGZus{}refGene    CDS     48267145        48267291        0.000000        \PYGZhy{}       0       gene\PYGZus{}id \PYGZdq{}NM\PYGZus{}001194986\PYGZdq{}; transcript\PYGZus{}id \PYGZdq{}NM\PYGZus{}001194986\PYGZdq{};
chr1   hg19\PYGZus{}refGene    exon    66081691        66081907        0.000000        +       .       gene\PYGZus{}id \PYGZdq{}NM\PYGZus{}002303\PYGZdq{}; transcript\PYGZus{}id \PYGZdq{}NM\PYGZus{}002303\PYGZdq{};
chr1   hg19\PYGZus{}refGene    CDS     145334684       145334792       0.000000        +       2       gene\PYGZus{}id \PYGZdq{}NM\PYGZus{}001039703\PYGZdq{}; transcript\PYGZus{}id \PYGZdq{}NM\PYGZus{}001039703\PYGZdq{};
chr1   hg19\PYGZus{}refGene    exon    172017752       172017890       0.000000        +       .       gene\PYGZus{}id \PYGZdq{}NM\PYGZus{}001136127\PYGZdq{}; transcript\PYGZus{}id \PYGZdq{}NM\PYGZus{}001136127\PYGZdq{};
chr1   hg19\PYGZus{}refGene    CDS     206589249       206589333       0.000000        +       2       gene\PYGZus{}id \PYGZdq{}NM\PYGZus{}001170637\PYGZdq{}; transcript\PYGZus{}id \PYGZdq{}NM\PYGZus{}001170637\PYGZdq{};
chr1   hg19\PYGZus{}refGene    exon    210573812       210574006       0.000000        +       .       gene\PYGZus{}id \PYGZdq{}NM\PYGZus{}001170580\PYGZdq{}; transcript\PYGZus{}id \PYGZdq{}NM\PYGZus{}001170580\PYGZdq{};
chr1   hg19\PYGZus{}refGene    CDS     235850249       235850347       0.000000        \PYGZhy{}       0       gene\PYGZus{}id \PYGZdq{}NM\PYGZus{}000081\PYGZdq{}; transcript\PYGZus{}id \PYGZdq{}NM\PYGZus{}000081\PYGZdq{};
chr1   hg19\PYGZus{}refGene    CDS     235880012       235880078       0.000000        \PYGZhy{}       1       gene\PYGZus{}id \PYGZdq{}NM\PYGZus{}000081\PYGZdq{}; transcript\PYGZus{}id \PYGZdq{}NM\PYGZus{}000081\PYGZdq{};
chr1   hg19\PYGZus{}refGene    exon    3417741 3417872 0.000000        \PYGZhy{}       .       gene\PYGZus{}id \PYGZdq{}NM\PYGZus{}001409\PYGZdq{}; transcript\PYGZus{}id \PYGZdq{}NM\PYGZus{}001409\PYGZdq{};
chr1   hg19\PYGZus{}refGene    exon    10190773        10190871        0.000000        +       .       gene\PYGZus{}id \PYGZdq{}NM\PYGZus{}006048\PYGZdq{}; transcript\PYGZus{}id \PYGZdq{}NM\PYGZus{}006048\PYGZdq{};

\PYGZdl{} head test.hg18.gtf
chr1   hg19\PYGZus{}refGene    CDS     48039732        48039878        0.000000        \PYGZhy{}       0       gene\PYGZus{}id \PYGZdq{}NM\PYGZus{}001194986\PYGZdq{}; transcript\PYGZus{}id \PYGZdq{}NM\PYGZus{}001194986\PYGZdq{};
chr1   hg19\PYGZus{}refGene    exon    65854279        65854495        0.000000        +       .       gene\PYGZus{}id \PYGZdq{}NM\PYGZus{}002303\PYGZdq{}; transcript\PYGZus{}id \PYGZdq{}NM\PYGZus{}002303\PYGZdq{};
chr1   hg19\PYGZus{}refGene    CDS     144046041       144046149       0.000000        +       2       gene\PYGZus{}id \PYGZdq{}NM\PYGZus{}001039703\PYGZdq{}; transcript\PYGZus{}id \PYGZdq{}NM\PYGZus{}001039703\PYGZdq{};
chr1   hg19\PYGZus{}refGene    exon    170284375       170284513       0.000000        +       .       gene\PYGZus{}id \PYGZdq{}NM\PYGZus{}001136127\PYGZdq{}; transcript\PYGZus{}id \PYGZdq{}NM\PYGZus{}001136127\PYGZdq{};
chr1   hg19\PYGZus{}refGene    CDS     204655872       204655956       0.000000        +       2       gene\PYGZus{}id \PYGZdq{}NM\PYGZus{}001170637\PYGZdq{}; transcript\PYGZus{}id \PYGZdq{}NM\PYGZus{}001170637\PYGZdq{};
chr1   hg19\PYGZus{}refGene    exon    208640435       208640629       0.000000        +       .       gene\PYGZus{}id \PYGZdq{}NM\PYGZus{}001170580\PYGZdq{}; transcript\PYGZus{}id \PYGZdq{}NM\PYGZus{}001170580\PYGZdq{};
chr1   hg19\PYGZus{}refGene    CDS     233916872       233916970       0.000000        \PYGZhy{}       0       gene\PYGZus{}id \PYGZdq{}NM\PYGZus{}000081\PYGZdq{}; transcript\PYGZus{}id \PYGZdq{}NM\PYGZus{}000081\PYGZdq{};
chr1   hg19\PYGZus{}refGene    CDS     233946635       233946701       0.000000        \PYGZhy{}       1       gene\PYGZus{}id \PYGZdq{}NM\PYGZus{}000081\PYGZdq{}; transcript\PYGZus{}id \PYGZdq{}NM\PYGZus{}000081\PYGZdq{};
chr1   hg19\PYGZus{}refGene    exon    3407601 3407732 0.000000        \PYGZhy{}       .       gene\PYGZus{}id \PYGZdq{}NM\PYGZus{}001409\PYGZdq{}; transcript\PYGZus{}id \PYGZdq{}NM\PYGZus{}001409\PYGZdq{};
chr1   hg19\PYGZus{}refGene    exon    10113360        10113458        0.000000        +       .       gene\PYGZus{}id \PYGZdq{}NM\PYGZus{}006048\PYGZdq{}; transcript\PYGZus{}id \PYGZdq{}NM\PYGZus{}006048\PYGZdq{};
\end{sphinxVerbatim}

NOTE:
\begin{enumerate}
\item {} 
Each feature  (exon, intron, UTR, etc) is processed separately and independently, and
we do NOT check if features originally belonging to the same gene were converted into the same gene.

\item {} 
If user want to liftover gene annotation files, use BED12 format.

\item {} 
If no output file was specified, output will be printed to screen (console). In this case, items failed to convert are also printed out.

\end{enumerate}


\section{Convert VCF format files}
\label{\detokenize{index:convert-vcf-format-files}}
\sphinxhref{http://www.1000genomes.org/wiki/Analysis/Variant\%20Call\%20Format/vcf-variant-call-format-version-41}{VCF}
(variant call format) is a flexible and extendable line-oriented text format developed by
the \sphinxhref{http://www.1000genomes.org/}{1000 Genome Project}. It is useful for representing single
nucleotide variants, indels, copy number variants, and structural variants. Chromosomes,
coordinates, and reference alleles are updated to a new assembly, and all the other fields
are not changed.

Typing command without any arguments will print help message:

\begin{sphinxVerbatim}[commandchars=\\\{\}]
\PYGZdl{} python2.7 CrossMap.py  vcf
\end{sphinxVerbatim}

Screen output:

\begin{sphinxVerbatim}[commandchars=\\\{\}]
\PYG{n}{usage}\PYG{p}{:}
  \PYG{n}{CrossMap}\PYG{o}{.}\PYG{n}{py} \PYG{n}{vcf} \PYG{n}{input\PYGZus{}chain\PYGZus{}file} \PYG{n}{input\PYGZus{}VCF\PYGZus{}file} \PYG{n}{ref\PYGZus{}genome\PYGZus{}file} \PYG{n}{output\PYGZus{}file}

\PYG{n}{Description}\PYG{p}{:}
  \PYG{l+s+s2}{\PYGZdq{}}\PYG{l+s+s2}{input\PYGZus{}chain\PYGZus{}file}\PYG{l+s+s2}{\PYGZdq{}} \PYG{o+ow}{and} \PYG{l+s+s2}{\PYGZdq{}}\PYG{l+s+s2}{input\PYGZus{}VCF\PYGZus{}file}\PYG{l+s+s2}{\PYGZdq{}} \PYG{n}{can} \PYG{n}{be} \PYG{n}{regular} \PYG{o+ow}{or} \PYG{n}{compressed} \PYG{p}{(}\PYG{o}{*}\PYG{o}{.}\PYG{n}{gz}\PYG{p}{,} \PYG{o}{*}\PYG{o}{.}\PYG{n}{Z}\PYG{p}{,}
  \PYG{o}{*}\PYG{o}{.}\PYG{n}{z}\PYG{p}{,} \PYG{o}{*}\PYG{o}{.}\PYG{n}{bz}\PYG{p}{,} \PYG{o}{*}\PYG{o}{.}\PYG{n}{bz2}\PYG{p}{,} \PYG{o}{*}\PYG{o}{.}\PYG{n}{bzip2}\PYG{p}{)} \PYG{n}{file}\PYG{p}{,} \PYG{n}{local} \PYG{n}{file} \PYG{o+ow}{or} \PYG{n}{URL} \PYG{p}{(}\PYG{n}{http}\PYG{p}{:}\PYG{o}{/}\PYG{o}{/}\PYG{p}{,} \PYG{n}{https}\PYG{p}{:}\PYG{o}{/}\PYG{o}{/}\PYG{p}{,} \PYG{n}{ftp}\PYG{p}{:}\PYG{o}{/}\PYG{o}{/}\PYG{p}{)}
  \PYG{n}{pointing} \PYG{n}{to} \PYG{n}{remote} \PYG{n}{file}\PYG{o}{.} \PYG{l+s+s2}{\PYGZdq{}}\PYG{l+s+s2}{ref\PYGZus{}genome\PYGZus{}file}\PYG{l+s+s2}{\PYGZdq{}} \PYG{o+ow}{is} \PYG{n}{genome} \PYG{n}{sequence} \PYG{n}{file} \PYG{n}{of} \PYG{l+s+s1}{\PYGZsq{}}\PYG{l+s+s1}{target}
  \PYG{n}{assembly}\PYG{l+s+s1}{\PYGZsq{}}\PYG{l+s+s1}{ in FASTA foramt.}

\PYG{n}{Example}\PYG{p}{:}
  \PYG{n}{CrossMap}\PYG{o}{.}\PYG{n}{py} \PYG{n}{vcf} \PYG{n}{hg19ToHg18}\PYG{o}{.}\PYG{n}{over}\PYG{o}{.}\PYG{n}{chain}\PYG{o}{.}\PYG{n}{gz} \PYG{n}{test}\PYG{o}{.}\PYG{n}{hg19}\PYG{o}{.}\PYG{n}{vcf} \PYG{n}{hg18}\PYG{o}{.}\PYG{n}{fa} \PYG{n}{test}\PYG{o}{.}\PYG{n}{hg18}\PYG{o}{.}\PYG{n}{vcf}
\end{sphinxVerbatim}

Example (Convert VCF file from hg19 to hg18):

\begin{sphinxVerbatim}[commandchars=\\\{\}]
\PYGZdl{} python CrossMap.py vcf hg19ToHg18.over.chain.gz test.hg19.vcf ../database/genome/hg18.fa  test.hg18.vcf
@ 2015\PYGZhy{}07\PYGZhy{}27 10:14:23: Read chain\PYGZus{}file:  ../data/hg19ToHg18.over.chain.gz
@ 2013\PYGZhy{}11\PYGZhy{}17 20:53:39: Creating index for ../database/genome/hg18.fa
@ 2015\PYGZhy{}07\PYGZhy{}27 10:14:50: Total entries: 497
@ 2015\PYGZhy{}07\PYGZhy{}27 10:14:50: Failed to map: 0

\PYGZdl{} grep \PYGZhy{}v \PYGZsq{}\PYGZsh{}\PYGZsq{} test.hg19.vcf  \textbar{}head \PYGZhy{}10
chr1   10933566        .       C       G       .       PASS    ADP=13;WT=0;HET=0;HOM=1;NC=0    GT:GQ:SDP:DP:RD:AD:FREQ:PVAL:RBQ:ABQ:RDF:RDR:ADF:ADR    1/1:7:13:13:0:13:100\PYGZpc{}:9.6148E\PYGZhy{}8:0:36:0:0:8:5
chr1   11187893        .       T       C       .       PASS    ADP=224;WT=0;HET=0;HOM=1;NC=0   GT:GQ:SDP:DP:RD:AD:FREQ:PVAL:RBQ:ABQ:RDF:RDR:ADF:ADR    1/1:133:226:224:0:224:100\PYGZpc{}:3.6518E\PYGZhy{}134:0:38:0:0:41:183
chr1   11205058        .       C       T       .       PASS    ADP=625;WT=0;HET=0;HOM=1;NC=0   GT:GQ:SDP:DP:RD:AD:FREQ:PVAL:RBQ:ABQ:RDF:RDR:ADF:ADR    1/1:255:643:625:0:625:100\PYGZpc{}:0E0:0:37:0:0:294:331
chr1   11292753        .       A       G       .       PASS    ADP=52;WT=0;HET=0;HOM=1;NC=0    GT:GQ:SDP:DP:RD:AD:FREQ:PVAL:RBQ:ABQ:RDF:RDR:ADF:ADR    1/1:27:52:52:2:50:96.15\PYGZpc{}:9.0394E\PYGZhy{}28:39:38:0:2:0:50
chr1   11318763        .       C       G       .       str10   ADP=88;WT=0;HET=0;HOM=1;NC=0    GT:GQ:SDP:DP:RD:AD:FREQ:PVAL:RBQ:ABQ:RDF:RDR:ADF:ADR    1/1:51:88:88:0:88:100\PYGZpc{}:1.7384E\PYGZhy{}52:0:38:0:0:1:87
chr1   11319587        .       A       G       .       PASS    ADP=70;WT=0;HET=0;HOM=1;NC=0    GT:GQ:SDP:DP:RD:AD:FREQ:PVAL:RBQ:ABQ:RDF:RDR:ADF:ADR    1/1:40:70:70:0:70:100\PYGZpc{}:1.0659E\PYGZhy{}41:0:38:0:0:0:70
chr1   16202995        .       C       T       .       PASS    ADP=463;WT=0;HET=1;HOM=0;NC=0   GT:GQ:SDP:DP:RD:AD:FREQ:PVAL:RBQ:ABQ:RDF:RDR:ADF:ADR    0/1:1:463:463:458:5:1.08\PYGZpc{}:3.0913E\PYGZhy{}2:37:33:188:270:4:1
chr1   27088546        .       A       T       .       PASS    ADP=124;WT=0;HET=1;HOM=0;NC=0   GT:GQ:SDP:DP:RD:AD:FREQ:PVAL:RBQ:ABQ:RDF:RDR:ADF:ADR    0/1:21:124:124:65:59:47.58\PYGZpc{}:1.7915E\PYGZhy{}22:37:38:59:6:55:4
chr1   27101390        .       T       C       .       str10   ADP=267;WT=0;HET=1;HOM=0;NC=0   GT:GQ:SDP:DP:RD:AD:FREQ:PVAL:RBQ:ABQ:RDF:RDR:ADF:ADR    0/1:1:267:267:262:5:1.87\PYGZpc{}:3.0665E\PYGZhy{}2:32:22:85:177:5:0
chr1   34007097        .       T       C       .       PASS    ADP=10;WT=0;HET=1;HOM=0;NC=0    GT:GQ:SDP:DP:RD:AD:FREQ:PVAL:RBQ:ABQ:RDF:RDR:ADF:ADR    0/1:1:10:10:6:4:40\PYGZpc{}:4.3344E\PYGZhy{}2:34:32:0:6:0:4

\PYGZdl{} grep \PYGZhy{}v \PYGZsq{}\PYGZsh{}\PYGZsq{} test.hg18.vcf  \textbar{}head \PYGZhy{}10
1      10856153        .       C       G       .       PASS    ADP=13;WT=0;HET=0;HOM=1;NC=0    GT:GQ:SDP:DP:RD:AD:FREQ:PVAL:RBQ:ABQ:RDF:RDR:ADF:ADR    1/1:7:13:13:0:13:100\PYGZpc{}:9.6148E\PYGZhy{}8:0:36:0:0:8:5
1      11110480        .       T       C       .       PASS    ADP=224;WT=0;HET=0;HOM=1;NC=0   GT:GQ:SDP:DP:RD:AD:FREQ:PVAL:RBQ:ABQ:RDF:RDR:ADF:ADR    1/1:133:226:224:0:224:100\PYGZpc{}:3.6518E\PYGZhy{}134:0:38:0:0:41:183
1      11127645        .       C       T       .       PASS    ADP=625;WT=0;HET=0;HOM=1;NC=0   GT:GQ:SDP:DP:RD:AD:FREQ:PVAL:RBQ:ABQ:RDF:RDR:ADF:ADR    1/1:255:643:625:0:625:100\PYGZpc{}:0E0:0:37:0:0:294:331
1      11215340        .       A       G       .       PASS    ADP=52;WT=0;HET=0;HOM=1;NC=0    GT:GQ:SDP:DP:RD:AD:FREQ:PVAL:RBQ:ABQ:RDF:RDR:ADF:ADR    1/1:27:52:52:2:50:96.15\PYGZpc{}:9.0394E\PYGZhy{}28:39:38:0:2:0:50
1      11241350        .       C       G       .       str10   ADP=88;WT=0;HET=0;HOM=1;NC=0    GT:GQ:SDP:DP:RD:AD:FREQ:PVAL:RBQ:ABQ:RDF:RDR:ADF:ADR    1/1:51:88:88:0:88:100\PYGZpc{}:1.7384E\PYGZhy{}52:0:38:0:0:1:87
1      11242174        .       A       G       .       PASS    ADP=70;WT=0;HET=0;HOM=1;NC=0    GT:GQ:SDP:DP:RD:AD:FREQ:PVAL:RBQ:ABQ:RDF:RDR:ADF:ADR    1/1:40:70:70:0:70:100\PYGZpc{}:1.0659E\PYGZhy{}41:0:38:0:0:0:70
1      16075582        .       C       T       .       PASS    ADP=463;WT=0;HET=1;HOM=0;NC=0   GT:GQ:SDP:DP:RD:AD:FREQ:PVAL:RBQ:ABQ:RDF:RDR:ADF:ADR    0/1:1:463:463:458:5:1.08\PYGZpc{}:3.0913E\PYGZhy{}2:37:33:188:270:4:1
1      26961133        .       A       T       .       PASS    ADP=124;WT=0;HET=1;HOM=0;NC=0   GT:GQ:SDP:DP:RD:AD:FREQ:PVAL:RBQ:ABQ:RDF:RDR:ADF:ADR    0/1:21:124:124:65:59:47.58\PYGZpc{}:1.7915E\PYGZhy{}22:37:38:59:6:55:4
1      26973977        .       T       C       .       str10   ADP=267;WT=0;HET=1;HOM=0;NC=0   GT:GQ:SDP:DP:RD:AD:FREQ:PVAL:RBQ:ABQ:RDF:RDR:ADF:ADR    0/1:1:267:267:262:5:1.87\PYGZpc{}:3.0665E\PYGZhy{}2:32:22:85:177:5:0
1      33779684        .       T       C       .       PASS    ADP=10;WT=0;HET=1;HOM=0;NC=0    GT:GQ:SDP:DP:RD:AD:FREQ:PVAL:RBQ:ABQ:RDF:RDR:ADF:ADR    0/1:1:10:10:6:4:40\PYGZpc{}:4.3344E\PYGZhy{}2:34:32:0:6:0:4

\PYGZdl{} grep \PYGZhy{}v \PYGZsq{}\PYGZsh{}\PYGZsq{} test.hg18.vcf.unmap      \PYGZsh{}coordinates are still based on hg19
chr14  20084444        .       G       C       .       PASS    ADP=253;WT=0;HET=1;HOM=0;NC=0   GT:GQ:SDP:DP:RD:AD:FREQ:PVAL:RBQ:ABQ:RDF:RDR:ADF:ADR    0/1:1:253:253:247:5:1.98\PYGZpc{}:3.0631E\PYGZhy{}2:38:39:123:124:5:0
chr14  20086290        .       T       C       .       PASS    ADP=441;WT=0;HET=1;HOM=0;NC=0   GT:GQ:SDP:DP:RD:AD:FREQ:PVAL:RBQ:ABQ:RDF:RDR:ADF:ADR    0/1:4:441:441:427:14:3.17\PYGZpc{}:5.4963E\PYGZhy{}5:37:38:236:191:6:8
\end{sphinxVerbatim}

NOTE:
\begin{enumerate}
\item {} 
Genome coordinates and reference allele will be updated to target assembly.

\item {} 
Reference genome is genome sequence of target assembly.

\item {} 
If the reference genome sequence file (../database/genome/hg18.fa) was not indexed, CrossMap will automatically indexed it (only the first time you run CrossMap).

\item {} 
Output files: \sphinxstyleemphasis{output\_file} and \sphinxstyleemphasis{output\_file.unmap}.

\item {} 
In the output VCF file, whether the chromosome IDs contain \sphinxquotedblleft{}chr\sphinxquotedblright{} or not depends on the format of the input VCF file.

\end{enumerate}


\chapter{Compare to UCSC liftover tool}
\label{\detokenize{index:compare-to-ucsc-liftover-tool}}
To access the accuracy of CrossMap, we randomly generated 10,000 genome intervals (download from \sphinxhref{https://sourceforge.net/projects/crossmap/files/hg19.rand.bed.gz/download}{here}) with the
fixed interval size of 200 bp from hg19. Then we converted them into hg18 using CrossMap
and \sphinxhref{http://genome.ucsc.edu/cgi-bin/hgLiftOver}{UCSC liftover tool} with default configurations. We compare CrossMap
to \sphinxhref{http://genome.ucsc.edu/cgi-bin/hgLiftOver}{UCSC liftover tool} because it is the most widely
used tool to convert genome coordinates.

CrossMap failed to convert 613 intervals, and UCSC liftover tool failed to convert 614
intervals. All failed intervals are exactly the same except one region (chr2 90542908 90543108).
UCSC failed to convert it because this region needs to be split twice:

\noindent\begin{tabulary}{\linewidth}{|L|L|L|}
\hline
\sphinxstylethead{\relax 
Original (hg19)
\unskip}\relax &\sphinxstylethead{\relax 
Split (hg19)
\unskip}\relax &\sphinxstylethead{\relax 
Target (hg18)
\unskip}\relax \\
\hline
chr2 90542908  90543108 -
&
chr2 90542908 90542933 -
&
chr2    89906445        89906470 -
\\
\hline
chr2 90542908  90543108 -
&
chr2 90542933 90543001 -
&
chr2    87414583        87414651 -
\\
\hline
chr2 90542908  90543108 -
&
chr2 90543010 90543108 -
&
chr2    87414276        87414374 -
\\
\hline\end{tabulary}


For genome intervals that were successfully converted to hg18, the start and end coordinates are
exactly the same between UCSC conversion and CrossMap conversion.

\noindent\scalebox{1.000000}{\sphinxincludegraphics[width=700\sphinxpxdimen,height=400\sphinxpxdimen]{{CrossMap_vs_UCSC}.png}}


\chapter{Citation}
\label{\detokenize{index:citation}}
Zhao, H., Sun, Z., Wang, J., Huang, H., Kocher, J.-P., \& Wang, L. (2013). CrossMap: a versatile tool for coordinate conversion between genome assemblies. Bioinformatics (Oxford, England), btt730.


\chapter{LICENSE}
\label{\detokenize{index:license}}
CrossMap is distributed under \sphinxhref{http://www.gnu.org/copyleft/gpl.html}{GNU General Public License}

This program is free software; you can redistribute it and/or
modify it under the terms of the GNU General Public License as
published by the Free Software Foundation; either version 2 of the
License, or (at your option) any later version. This program is distributed in the hope that it will be useful,
but WITHOUT ANY WARRANTY; without even the implied warranty of
MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
General Public License for more details. You should have received a copy of the GNU General Public License
along with this program; if not, write to the Free Software
Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA
02110-1301 USA


\chapter{Contact}
\label{\detokenize{index:contact}}\begin{itemize}
\item {} 
Wang.Liguo AT mayo.edu

\end{itemize}

\noindent\scalebox{1.000000}{\sphinxincludegraphics[width=80\sphinxpxdimen,height=80\sphinxpxdimen]{{mayo}.jpg}}

\noindent\scalebox{1.000000}{\sphinxincludegraphics[width=150\sphinxpxdimen,height=80\sphinxpxdimen]{{mdacc}.jpg}}

\noindent\scalebox{1.000000}{\sphinxincludegraphics[width=150\sphinxpxdimen,height=100\sphinxpxdimen]{{sourceforge}.jpg}}



\renewcommand{\indexname}{Index}
\printindex
\end{document}