
<!DOCTYPE html>

<html xmlns="http://www.w3.org/1999/xhtml">
  <head>
    <meta charset="utf-8" />
    <title>What is CrossMap ? &#8212; CrossMap 0.4.0 documentation</title>
    <link rel="stylesheet" href="../_static/nature.css" type="text/css" />
    <link rel="stylesheet" href="../_static/pygments.css" type="text/css" />
    <script type="text/javascript" id="documentation_options" data-url_root="../" src="../_static/documentation_options.js"></script>
    <script type="text/javascript" src="../_static/jquery.js"></script>
    <script type="text/javascript" src="../_static/underscore.js"></script>
    <script type="text/javascript" src="../_static/doctools.js"></script>
    <script type="text/javascript" src="../_static/language_data.js"></script>
    <link rel="index" title="Index" href="../genindex.html" />
    <link rel="search" title="Search" href="../search.html" /> 
  </head><body>
    <div class="related" role="navigation" aria-label="related navigation">
      <h3>Navigation</h3>
      <ul>
        <li class="right" style="margin-right: 10px">
          <a href="../genindex.html" title="General Index"
             accesskey="I">index</a></li>
        <li class="nav-item nav-item-0"><a href="../index.html">CrossMap 0.4.0 documentation</a> &#187;</li> 
      </ul>
    </div>  

    <div class="document">
      <div class="documentwrapper">
        <div class="bodywrapper">
          <div class="body" role="main">
            
  <div class="toctree-wrapper compound">
</div>
<a class="reference internal image-reference" href="backup/_static/logo.png"><img alt="alternate text" src="backup/_static/logo.png" style="width: 375.0px; height: 75.0px;" /></a>
<div class="section" id="what-is-crossmap">
<h1>What is CrossMap ?<a class="headerlink" href="#what-is-crossmap" title="Permalink to this headline">¶</a></h1>
<p>CrossMap is a program for genome coordinates conversion between <em>different assemblies</em>
(such as <a class="reference external" href="http://www.ncbi.nlm.nih.gov/assembly/2928/">hg18 (NCBI36)</a>   &lt;=&gt; <a class="reference external" href="http://www.ncbi.nlm.nih.gov/assembly/2758/">hg19 (GRCh37)</a>).
It supports commonly used file formats including BAM, CRAM, SAM, Wiggle, BigWig, BED, GFF, GTF and VCF.</p>
</div>
<div class="section" id="how-crossmap-works">
<h1>How CrossMap works?<a class="headerlink" href="#how-crossmap-works" title="Permalink to this headline">¶</a></h1>
<a class="reference internal image-reference" href="backup/_static/howitworks.png"><img alt="alternate text" src="backup/_static/howitworks.png" style="width: 510.0px; height: 212.5px;" /></a>
</div>
<div class="section" id="release-history">
<h1>Release history<a class="headerlink" href="#release-history" title="Permalink to this headline">¶</a></h1>
<ul class="simple">
<li><p>10/09/2019: Release version 0.3.8.</p></li>
</ul>
<blockquote>
<div><ul class="simple">
<li><p>The University of California holds the copyrights in the UCSC chain files. As requested by UCSC, all UCSC generated chain files will be permanently removed from this website and the CrossMap distributions.</p></li>
</ul>
</div></blockquote>
<ul class="simple">
<li><p>07/22/2019: Release version 0.3.6.</p>
<ul>
<li><p>Support MAF (mutation annotation format).</p></li>
<li><p>Fix error “TypeError: AlignmentHeader does not support item assignment (use header.to_dict()” when lifting over BAM files. User does not need to downgrade pysam to 0.13.0 to lift over BAM files.</p></li>
</ul>
</li>
<li><p>04/01/2019: Release version 0.3.4.</p></li>
</ul>
<blockquote>
<div><ul class="simple">
<li><p>Fix bugs when chromosome IDs (of the source genome) in chain file do not have ‘chr’ prefix (such as “GRCh37ToHg19.over.chain.gz”). This version also allows CrossMap to detct if a VCF mapping was inverted, and if so reverse complements the altenerative allele (Thanks Andrew Yates). Improve wording.</p></li>
</ul>
</div></blockquote>
<ul class="simple">
<li><p>01/07/2019: Release version 0.3.3.</p></li>
</ul>
<blockquote>
<div><ul class="simple">
<li><p>Version 0.3.3 is exactly the same as Version 0.3.2. The reason to release this version is that CrossMap-0.3.2.tar.gz was broken when uploading to pypi.</p></li>
</ul>
</div></blockquote>
<ul class="simple">
<li><p>12/14/18: Release version 0.3.2.</p></li>
</ul>
<blockquote>
<div><ul class="simple">
<li><p>Fix the key error problem (e.g  <em>KeyError: “sequence ‘b‘7_KI270803v1_alt’’ not present”</em>). This error happens when a locus from the orignal assembly is mapped to a “alternative”, “unplaced” or “unlocalized” contig in the target assembly, and this “target contig” does not exist in your target_ref.fa. In version 0.3.2, such loci will be silently skipped and saved to the “.unmap” file.</p></li>
</ul>
</div></blockquote>
<ul class="simple">
<li><p>11/05/18: Release version 0.3.0:</p></li>
</ul>
<blockquote>
<div><ul class="simple">
<li><p>v0.3.0 or newer will Support Python3. Previous versions support Python2.7.*</p></li>
<li><p>add <a class="reference external" href="https://github.com/deeptools/pyBigWig">pyBigWig</a> as a dependency.</p></li>
</ul>
</div></blockquote>
<ul class="simple">
<li><p>09/06/17: Release version 0.2.8:</p></li>
</ul>
<blockquote>
<div><ul class="simple">
<li><p>In Bam file lift over: fixed the bug “CrossMap does not set the unmapped read flag for the first read in pair when it is unmapped”.</p></li>
<li><p>In VCF file lift over: Update the “contig field” in VCF header section. Contig name and size will be changed from the old assembly to the new assembly.</p></li>
</ul>
</div></blockquote>
<ul class="simple">
<li><p>09/06/17: Release version 0.2.7:</p></li>
</ul>
<blockquote>
<div><ul class="simple">
<li><p>In VCF file lift over: fixed the bug “non-standard chromosome IDs were not converted”.</p></li>
</ul>
</div></blockquote>
<ul class="simple">
<li><p>05/09/17: Release version 0.2.6:</p></li>
</ul>
<blockquote>
<div><ul class="simple">
<li><p>In BAM file lift over: fixed bugs during BAM file sorting and indexing steps (works with pysam v0.11.1).</p></li>
<li><p>In BAM file lift over: fixed bugs “the read group type is automatically and wrongly changed from Z to A” (<a class="reference external" href="https://github.com/pysam-developers/pysam/issues/113">https://github.com/pysam-developers/pysam/issues/113</a>).</p></li>
</ul>
</div></blockquote>
<ul class="simple">
<li><p>10/7/16: Release version 0.2.5:</p></li>
</ul>
<blockquote>
<div><ul class="simple">
<li><p>fixed bugs during single-end BAM file conversion.</p></li>
<li><p>Add optional tags to the output BAM file. Details see: <a class="reference internal" href="../index.html#bam-conversion"><span class="std std-ref">Convert BAM/CRAM/SAM format files</span></a>.</p></li>
</ul>
</div></blockquote>
<ul class="simple">
<li><p>08/18/16: Release version 0.2.4:</p></li>
</ul>
<blockquote>
<div><ul class="simple">
<li><p>fixed bugs during BAM file conversion:</p></li>
</ul>
<blockquote>
<div><ul class="simple">
<li><p>When the strand of the ead changes, the seq filed is reverse complemented and the quality field is reversed.</p></li>
</ul>
</div></blockquote>
<ul class="simple">
<li><p>In the output VCF file, if the reference allele field is empty:</p></li>
</ul>
<blockquote>
<div><ul class="simple">
<li><p>Use CrossMap v0.2.4. Update pysam to the latest version. And make sure chromosome IDs in the reference genome file are in the form of “chr1”, “chr2”, …, “chrX”,”chrY” (but not “1”, “2”, …, “X”,”Y”, in this case, pysam cannot index your reference genome file for some unknown reasons.).</p></li>
</ul>
</div></blockquote>
<ul class="simple">
<li><p>to upgrade, run: <strong>pip install CrossMap –upgrade</strong></p></li>
</ul>
</div></blockquote>
<ul class="simple">
<li><p>04/13/16: Release version 0.2.3:</p></li>
</ul>
<blockquote>
<div><ul class="simple">
<li><p>Same as v0.2.2.</p></li>
<li><p>Two dependency packages bx-python and pysam do not ship with CrossMap starting from v0.2.3 .</p></li>
<li><p>Users could install CrossMap using pip: <strong>pip install CrossMap</strong>. Note: bx-python and pysam will be installed automatically if they haven’t been installed before.</p></li>
</ul>
</div></blockquote>
<ul class="simple">
<li><p>11/10/15: Release version 0.2.2: Generate *.unmap files (regions that cannot be unambiguously converted) when converting BED, GTF, GFF files. This version also supports genePred (bed12+8) format. (Thanks for Andrew Yates from EMBL-EBI)</p></li>
<li><p>08/26/15: Release version 0.2.1: Very minor change, same as 0.2.</p></li>
<li><p>08/11/15: Release version 0.2: Fixed the bug that CrossMap will not convert wiggle format files due to name collision with bx python.</p></li>
<li><p>07/27/15: Release version 0.1.9. For VCF file conversion in v0.1.9:</p></li>
</ul>
<blockquote>
<div><ul class="simple">
<li><p>CrossMap uses the indexed reference genome (target assembly) sequences rather than load the entire file into memory. Users could index their reference genome file using <strong>samtools faidx</strong> before running CrossMap, otherwise, CrossMap will index it automatically the first time you run it.</p></li>
<li><p>In the output VCF file, whether the chromosome IDs contain “chr” or not depends on the input format.</p></li>
</ul>
</div></blockquote>
<ul class="simple">
<li><p>05/15/15: Release version 0.1.8: Fixed the bug that CrossMap will output invalid VCF file when the input VCF file contains a INFO field with whitespace.</p></li>
<li><p>05/04/15: Release version 0.1.7: Address the problem that CrossMap does not convert strand in inversions when input file is BED6 or BED12 format.</p></li>
<li><p>11/06/14: Release version 0.1.6: Fixed “negative coordinates” bug.</p></li>
<li><p>08/05/14: Release version 0.1.5: Support compressed (*.gz, *.Z, *.z, *.bz, *.bz2, *.bzip2) wiggle file as input.</p></li>
<li><p>05/19/14: add chain files for hg38-&gt;hg19, hg19-&gt;hg38, hg18-&gt;hg38, hg19-&gt;GRCh37, GRCh37-&gt;hg19. In CrossMap v0.1.4, conversion results of BAM/SAM files can be directed to STDOUT to support piping.</p></li>
<li><p>12/12/13: CrossMap was accepted by <a class="reference external" href="http://bioinformatics.oxfordjournals.org/content/early/2013/12/18/bioinformatics.btt730.short?rss=1">Bioinformatics</a></p></li>
<li><p>10/23/13: CrossMap (0.1.3) was released</p></li>
</ul>
</div>
<div class="section" id="installation">
<h1>Installation<a class="headerlink" href="#installation" title="Permalink to this headline">¶</a></h1>
<div class="section" id="use-pip-to-install-crossmap">
<h2>Use pip to install CrossMap<a class="headerlink" href="#use-pip-to-install-crossmap" title="Permalink to this headline">¶</a></h2>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="n">pip3</span> <span class="n">install</span> <span class="n">CrossMap</span>  <span class="c1">#Install CrossMap supporting Python3</span>
<span class="n">pip2</span> <span class="n">install</span> <span class="n">CrossMap</span>  <span class="c1">#Install CrossMap supporting Python2.7.*</span>
</pre></div>
</div>
</div>
<div class="section" id="use-pip-to-upgrade-crossmap">
<h2>Use pip to upgrade CrossMap<a class="headerlink" href="#use-pip-to-upgrade-crossmap" title="Permalink to this headline">¶</a></h2>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="n">pip3</span> <span class="n">install</span> <span class="n">CrossMap</span> <span class="o">--</span><span class="n">upgrade</span>        <span class="c1">#upgrade CrossMap supporting Python3</span>
<span class="n">pip2</span> <span class="n">install</span> <span class="n">CrossMap</span> <span class="o">--</span><span class="n">upgrade</span>        <span class="c1">#upgrade CrossMap supporting Python2.7.*</span>
</pre></div>
</div>
</div>
</div>
<div class="section" id="input-and-output">
<h1>Input and Output<a class="headerlink" href="#input-and-output" title="Permalink to this headline">¶</a></h1>
<p>CrossMap basically needs 2 input files.  <a class="reference external" href="http://genome.ucsc.edu/goldenPath/help/chain.html">chain</a>
format file describing genome-wide pairwise alignments between assemblies and the file containing
genome coordinates that you want to convert to different assembly. If input file is in VCF
format, a reference genome sequence file(in FASTA format) is needed.</p>
<div class="section" id="chain-file">
<h2>Chain file<a class="headerlink" href="#chain-file" title="Permalink to this headline">¶</a></h2>
<ol class="arabic simple">
<li><p><strong>Download Ensembl chain files</strong></p></li>
</ol>
<blockquote>
<div><ul class="simple">
<li><p>Human to Human: <a class="reference external" href="ftp://ftp.ensembl.org/pub/assembly_mapping/homo_sapiens/">ftp://ftp.ensembl.org/pub/assembly_mapping/homo_sapiens/</a></p></li>
<li><p>Mouse to Mouse: <a class="reference external" href="ftp://ftp.ensembl.org/pub/assembly_mapping/mus_musculus/">ftp://ftp.ensembl.org/pub/assembly_mapping/mus_musculus/</a></p></li>
<li><p>Other organisms: <a class="reference external" href="ftp://ftp.ensembl.org/pub/assembly_mapping/">ftp://ftp.ensembl.org/pub/assembly_mapping/</a></p></li>
</ul>
</div></blockquote>
<ol class="arabic simple" start="2">
<li><p><strong>Download UCSC chain files</strong></p></li>
</ol>
<blockquote>
<div><ul class="simple">
<li><p>Chain files from hg38 (GRCh38) to hg19 and all other organisms: <a class="reference external" href="http://hgdownload.soe.ucsc.edu/goldenPath/hg38/liftOver/">http://hgdownload.soe.ucsc.edu/goldenPath/hg38/liftOver/</a></p></li>
<li><p>Chain File from hg19 (GRCh37) to hg17/hg18/hg38 and all other organisms: <a class="reference external" href="http://hgdownload.soe.ucsc.edu/goldenPath/hg19/liftOver/">http://hgdownload.soe.ucsc.edu/goldenPath/hg19/liftOver/</a></p></li>
<li><p>Chain File from mm10 (GRCm38) to mm9 and all other organisms: <a class="reference external" href="http://hgdownload.soe.ucsc.edu/goldenPath/mm10/liftOver/">http://hgdownload.soe.ucsc.edu/goldenPath/mm10/liftOver/</a></p></li>
</ul>
</div></blockquote>
</div>
<div class="section" id="user-input-file">
<h2>User Input file<a class="headerlink" href="#user-input-file" title="Permalink to this headline">¶</a></h2>
<ol class="arabic simple">
<li><p><a class="reference external" href="http://samtools.sourceforge.net/SAMv1.pdf">BAM</a>, <a class="reference external" href="https://samtools.github.io/hts-specs/CRAMv3.pdf">CRAM</a>, or <a class="reference external" href="http://samtools.sourceforge.net/SAMv1.pdf/">SAM</a> format.</p></li>
<li><p><a class="reference external" href="http://genome.ucsc.edu/FAQ/FAQformat.html#format1">BED</a> or BED-like format. BED file must has at least 3 columns (‘chrom’, ‘start’, ‘end’).</p></li>
<li><p><a class="reference external" href="http://genome.ucsc.edu/goldenPath/help/wiggle.html">Wiggle</a> format. “variableStep”, “fixedStep” and “bedGraph” wiggle line are supported.</p></li>
<li><p><a class="reference external" href="http://genome.ucsc.edu/goldenPath/help/bigWig.html">BigWig</a> format.</p></li>
<li><p><a class="reference external" href="http://genome.ucsc.edu/FAQ/FAQformat.html#format3">GFF</a> or <a class="reference external" href="http://genome.ucsc.edu/FAQ/FAQformat.html#format4">GTF</a> format.</p></li>
<li><p><a class="reference external" href="http://vcftools.sourceforge.net/index.html">VCF</a> format.</p></li>
</ol>
</div>
<div class="section" id="output-file">
<h2>Output file<a class="headerlink" href="#output-file" title="Permalink to this headline">¶</a></h2>
<p>Format of Output files depends on the input format</p>
<table class="docutils align-default">
<colgroup>
<col style="width: 13%" />
<col style="width: 87%" />
</colgroup>
<thead>
<tr class="row-odd"><th class="head"><p>Input_format</p></th>
<th class="head"><p>Output_format</p></th>
</tr>
</thead>
<tbody>
<tr class="row-even"><td><p>BED</p></td>
<td><p>BED (Genome coordinates will be updated to the target assembly)</p></td>
</tr>
<tr class="row-odd"><td><p>BAM</p></td>
<td><p>BAM (Genome coordinates, header section, all SAM flags, insert size will be updated accordingly)</p></td>
</tr>
<tr class="row-even"><td><p>CRAM</p></td>
<td><p>BAM (require pysam &gt;= 0.8.2)</p></td>
</tr>
<tr class="row-odd"><td><p>SAM</p></td>
<td><p>SAM (Genome coordinates, header section, all SAM flags, insert size will be updated accordingly)</p></td>
</tr>
<tr class="row-even"><td><p>Wiggle</p></td>
<td><p>BigWig</p></td>
</tr>
<tr class="row-odd"><td><p>BigWig</p></td>
<td><p>BigWig</p></td>
</tr>
<tr class="row-even"><td><p>GFF</p></td>
<td><p>GFF (Genome coordinates will be updated to the target assembly)</p></td>
</tr>
<tr class="row-odd"><td><p>GTF</p></td>
<td><p>GTF (Genome coordinates will be updated to the target assembly)</p></td>
</tr>
<tr class="row-even"><td><p>VCF</p></td>
<td><p>VCF (Genome coordinates and reference alleles will be updated to the target assembly)</p></td>
</tr>
</tbody>
</table>
</div>
</div>
<div class="section" id="usage">
<h1>Usage<a class="headerlink" href="#usage" title="Permalink to this headline">¶</a></h1>
<p>Run CrossMap.py <strong>without</strong> any arguments will print help message:</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span># run CrossMap without argument
$ python CrossMap.py
</pre></div>
</div>
<p>Program: CrossMap (v0.3.4)</p>
<blockquote>
<div><dl class="simple">
<dt>Description:</dt><dd><p>CrossMap is a program for convenient conversion of genome coordinates between
assembly versions (e.g. from human hg18 to hg19 or vice versa). It supports file
in BAM, CRAM, SAM, BED, Wiggle, BigWig, GFF, GTF and VCF format.</p>
</dd>
</dl>
<p>Usage: CrossMap.py &lt;command&gt; [options]</p>
<blockquote>
<div><p>bam  convert alignment file in BAM, CRAM or SAM format.
bed  convert genome coordinate or annotation file in BED, bedGraph or other BED-like format.
bigwig       convert genome coordinate file in BigWig format.
gff  convert genome coordinate or annotation file in GFF or GTF format.
vcf  convert genome coordinate file in VCF format.
wig  convert genome coordinate file in Wiggle, or bedGraph format.</p>
</div></blockquote>
</div></blockquote>
<p>Run CrossMap.py <strong>with</strong> a command keyword will print help message for the command. For example:</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span>$ python CrossMap.py bed

#Screen output

Usage
-----
  CrossMap.py bed chain_file input_bed_file [output_file]

Description
-----------
  Convert BED format file. The &quot;chain_file&quot; and &quot;input_bed_file&quot; can be regular or
  compressed (*.gz, *.Z, *.z, *.bz, *.bz2, *.bzip2) file, local file or URL
  (http://, https://, ftp://) pointing to remote file. BED format file must have
  at least 3 columns (chrom, start, end). If  no &quot;output_file&quot; is specified,
  output will be directed to the screen (console).

Example1 (write output to file)
-------------------------------
  CrossMap.py bed hg18ToHg19.over.chain.gz test.hg18.bed test.hg19.bed

Example2 (write output to screen)
---------------------------------
  CrossMap.py bed hg18ToHg19.over.chain.gz test.hg18.bed
</pre></div>
</div>
<div class="section" id="convert-bed-format-files">
<h2>Convert BED format files<a class="headerlink" href="#convert-bed-format-files" title="Permalink to this headline">¶</a></h2>
<p>A <a class="reference external" href="http://genome.ucsc.edu/FAQ/FAQformat.html#format1">BED</a> (Browser Extensible Data) file
is a tab-delimited text file describing genome regions or gene annotations.
It consists of one line per feature, each containing 3-12 columns.
CrossMap converts BED files with less than 12 columns to a different assembly by updating the
chromosome and genome coordinates only; all other columns remain unchanged. Regions from old
assembly mapping to multiple locations to the new assembly will be split.  For 12-columns BED
files, all columns will be updated accordingly except the 4th column (name of bed line), 5th
column (score value) and 9th column (RGB value describing the display color). 12-column BED
files usually define multiple blocks (e.g. exon); if any of the exons fails to map to a new
assembly, the whole BED line is skipped.</p>
<p>The input BED file can be plain text file, compressed file with extension of .gz, .Z, .z,
.bz, .bz2 and .bzip2, or even a URL pointing to accessible remote files (<a class="reference external" href="http://">http://</a>, <a class="reference external" href="https://">https://</a>
and <a class="reference external" href="ftp://">ftp://</a>). Compressed remote files are not supported. The output is a BED format file with
exact the same number of columns as the original one.</p>
<p>Standard <a class="reference external" href="http://genome.ucsc.edu/FAQ/FAQformat.html#format1">BED</a> format has 12 columns, but CrossMap also supports BED-like formats:</p>
<ul class="simple">
<li><p>BED3: The first 3 columns (“chrom”, “start”, “end”) of BED format file.</p></li>
<li><p>BED6: The first 6 columns (“chrom”, “start”, “end”, “name”, “score”, “strand”) of BED format file.</p></li>
<li><p>Other: Format has at least 3 columns (“chrom”, “start”, “end”) and no more than 12 columns. All other columns are arbitrary.</p></li>
</ul>
<p>NOTE:</p>
<ol class="arabic simple">
<li><p>For BED-like formats mentioned above, CrossMap only updates “chrom (1st column)”, “start (2nd column) “, “end (3rd column) ” and “strand” (if any). All other columns will keep AS-IS.</p></li>
<li><p>Lines starting with ‘#’, ‘browser’, ‘track’ will be skipped.</p></li>
<li><p>Lines will less than 3 columns will be skipped.</p></li>
<li><p>2nd-column and 3-column must be integers, otherwise, the line will be skipped.</p></li>
<li><p>“+” strand is assumed if no strand information was found.</p></li>
<li><p>For standard BED format (12 columns). If any of the defined exon blocks cannot be uniquely mapped to target assembly, the whole entry will be skipped.</p></li>
<li><p>“input_chain_file” and “input_bed_file” can be regular or compressed (.gz, .Z, .z, .bz, .bz2, .bzip2) file, local file or URL (<a class="reference external" href="http://">http://</a>, <a class="reference external" href="https://">https://</a>, <a class="reference external" href="ftp://">ftp://</a>) pointing to remote file.</p></li>
<li><p>If output_file was not specified, results will be printed to screen (console). In this case, the original bed entries (include items failed to convert) were also printed out.</p></li>
<li><p>If input region cannot be consecutively mapped target assembly, it will be split.</p></li>
<li><p>*.unmap file contains regions that cannot be unambiguously converted.</p></li>
</ol>
<p>Example (run CrossMap with <strong>no</strong> <em>output_file</em> specified):</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span>$ python CrossMap.py bed hg18ToHg19.over.chain.gz test.hg18.bed3

# Conversion results were printed to screen directly (column1-3 are hg18 based, column5-7 are hg19 based)::
chr1   142614848       142617697       -&gt;      chr1    143903503       143906352
chr1   142617697       142623312       -&gt;      chr1    143906355       143911970
chr1   142623313       142623350       -&gt;      chr1    143911971       143912008
</pre></div>
</div>
<p>Example (run CrossMap with <em>output_file</em> <strong>(test.hg19.bed3)</strong> specified):</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span>$ python CrossMap.py bed hg18ToHg19.over.chain.gz test.hg18.bed3 test.hg19.bed3

$ cat test.hg19.bed3
chr1   143903503       143906352
chr1   143906355       143911970
chr1   143911971       143912008
</pre></div>
</div>
<p>Example (one input region was split because it cannot be consecutively mapped target assembly):</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span>$ python CrossMap.py bed hg18ToHg19.over.chain.gz test.hg18.bed3

chr10  81346644        81349952        +       -&gt;      chr10   81356692        81360000        +
chr10  81349952        81364937        +       -&gt;      chr10   81360000        81374985        +
chr10  81364952        81365854        +       -&gt;      chr10   81375000        81375902        +
chr10  81365875        81369946        +       -&gt;      chr10   81375929        81380000        +
chr10  81369946        81370453        +       -&gt;      chr10   81380000        81380507        +
chr10  81370483        81371363        +       -&gt;      chr10   81380539        81381419        +
chr10  81371363        81371365        +       -&gt;      chr10   62961832        62961834        +
chr10  81371412        81371432        +       (split.1:chr10:81371412:81371422:+)     chr10   62961775        62961785        +
chr10  81371412        81371432        +       (split.2:chr10:81371422:81371432:+)     chrX    63278348        63278358        +
</pre></div>
</div>
<p>Example (Use <strong>bed</strong> command to convert a bedGraph file, output another bedGraph file. If Use <strong>wig</strong> command to convert a bedGraph file, output a <strong>bigWig</strong> file. ):</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span>$ python3 CrossMap.py bed hg19ToHg38.over.chain.gz 4_hg19.bgr

chrX   5873316 5873391 2.0     -&gt;      chrX    5955275 5955350 2.0
chrX   5873673 5873710 0.8     -&gt;      chrX    5955632 5955669 0.8
chrX   5873710 5873785 1.4     -&gt;      chrX    5955669 5955744 1.4
chrX   5873896 5873929 0.9     -&gt;      chrX    5955855 5955888 0.9
chrX   5873929 5874004 1.5     -&gt;      chrX    5955888 5955963 1.5
chrX   5874230 5874471 0.3     -&gt;      chrX    5956189 5956430 0.3
chrX   5874471 5874518 0.9     -&gt;      chrX    5956430 5956477 0.9

$ python3 CrossMap.py wig hg19ToHg38.over.chain.gz 4_hg19.bgr output_hg38
@ 2018-11-06 00:09:11: Read chain_file:  hg19ToHg38.over.chain.gz
@ 2018-11-06 00:09:12: Liftover wiggle file: 4_hg19.bgr ==&gt; output_hg38.bgr
@ 2018-11-06 00:09:12: Merging overlapped entries in bedGraph file ...
@ 2018-11-06 00:09:12: Sorting bedGraph file:output_hg38.bgr
@ 2018-11-06 00:09:12: Writing header to &quot;output_hg38.bw&quot; ...
@ 2018-11-06 00:09:12: Writing entries to &quot;output_hg38.bw&quot; ...
</pre></div>
</div>
</div>
<div class="section" id="convert-bam-cram-sam-format-files">
<span id="bam-conversion"></span><h2>Convert BAM/CRAM/SAM format files<a class="headerlink" href="#convert-bam-cram-sam-format-files" title="Permalink to this headline">¶</a></h2>
<p><a class="reference external" href="http://samtools.sourceforge.net/samtools.shtml#5">SAM</a> (Sequence Alignment Map) format
is a generic format for storing sequencing alignments, and BAM is binary and compressed
version of SAM (<a class="reference external" href="http://bioinformatics.oxfordjournals.org/content/25/16/2078.full">Li et al., 2009</a>).
<a class="reference external" href="https://samtools.github.io/hts-specs/CRAMv3.pdf">CRAM</a> was designed to be an efficient reference-based
alternative to the <a class="reference external" href="http://samtools.sourceforge.net/samtools.shtml#5">SAM</a> and BAM file formats
Most high-throughput sequencing  (HTS) alignments were in SAM/BAM format and many HTS analysis
tools work with SAM/BAM format. CrossMap updates chromosomes, genome coordinates, header
sections, and all SAM flags accordingly.  CrossMap’s version number is inserted into
the header section, along with the names of the original BAM file and the chain file.  For
pair-end sequencing, insert size is also recalculated. The input BAM file should be sorted
and indexed properly using samTools (<a class="reference external" href="http://bioinformatics.oxfordjournals.org/content/25/16/2078.full">Li et al., 2009</a>).
Output format is determined from the input format and BAM output will be sorted and indexed automatically.</p>
<p>Typing command without any arguments will print help message:</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span>$ python CrossMap.py bam

Usage: CrossMap.py bam chain_file input_file output_file [options]
Note: If output_file == STDOUT or -, CrossMap will write BAM file to the screen

Options:
  -m INSERT_SIZE, --mean=INSERT_SIZE
                        Average insert size of pair-end sequencing (bp).
                        [default=200.0]
  -s INSERT_SIZE_STDEV, --stdev=INSERT_SIZE_STDEV
                        Stanadard deviation of insert size. [default=30.0]
  -t INSERT_SIZE_FOLD, --times=INSERT_SIZE_FOLD
                        A mapped pair is considered as &quot;proper pair&quot; if both
                        ends mapped to different strand and the distance
                        between them is less then &#39;-t&#39; * stdev from the mean.
                        [default=3.0]
  -a, --append-tags     Add tag to each alignment.
</pre></div>
</div>
<p>Example (Convert BAM from hg19 to hg18):</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span># add optional tags using &#39;-a&#39; (recommend always use &#39;-a&#39; option)

$ CrossMap.py bam -a ../data/hg19ToHg18.over.chain.gz test.hg19.bam test.hg18
Insert size = 200.000000
Insert size stdev = 30.000000
Number of stdev from the mean = 3.000000
Add tags to each alignment = True
@ 2016-10-07 15:29:06: Read chain_file:  ../data/hg19ToHg18.over.chain.gz
@ 2016-10-07 15:29:07: Liftover BAM file: test.hg19.bam ==&gt; test.hg18.bam
@ 2016-10-07 15:29:14: Done!
@ 2016-10-07 15:29:14: Sort &quot;test.hg18.bam&quot; ...
@ 2016-10-07 15:29:15: Index &quot;test.hg18.sorted.bam&quot; ...
Total alignments:99914
       QC failed: 0
       R1 unique, R2 unique (UU): 96094
       R1 unique, R2 unmapp (UN): 3579
       R1 unique, R2 multiple (UM): 0
       R1 multiple, R2 multiple (MM): 0
       R1 multiple, R2 unique (MU): 233
       R1 multiple, R2 unmapped (MN): 8
       R1 unmap, R2 unmap (NN): 0
       R1 unmap, R2 unique (NU): 0
       R1 unmap, R2 multiple (NM): 0
</pre></div>
</div>
<p># BAM/SAM header sections was updated:</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span>$ samtools view -H  test.hg19.bam
@SQ    SN:chr1 LN:249250621
@SQ    SN:chr2 LN:243199373
@SQ    SN:chr3 LN:198022430
@SQ    SN:chr4 LN:191154276
@SQ    SN:chr5 LN:180915260
@SQ    SN:chr6 LN:171115067
@SQ    SN:chr7 LN:159138663
@SQ    SN:chr8 LN:146364022
@SQ    SN:chr9 LN:141213431
@SQ    SN:chr10        LN:135534747
@SQ    SN:chr11        LN:135006516
@SQ    SN:chr12        LN:133851895
@SQ    SN:chr13        LN:115169878
@SQ    SN:chr14        LN:107349540
@SQ    SN:chr15        LN:102531392
@SQ    SN:chr16        LN:90354753
@SQ    SN:chr17        LN:81195210
@SQ    SN:chr18        LN:78077248
@SQ    SN:chr19        LN:59128983
@SQ    SN:chr20        LN:63025520
@SQ    SN:chr21        LN:48129895
@SQ    SN:chr22        LN:51304566
@SQ    SN:chrX LN:155270560
@SQ    SN:chrY LN:59373566
@SQ    SN:chrM LN:16571
@RG    ID:Sample_618545BE      SM:Sample_618545BE      LB:Sample_618545BE      PL:Illumina
@PG    ID:bwa  PN:bwa  VN:0.6.2-r126

$ samtools view -H  test.hg18.bam
@HD    VN:1.0  SO:coordinate
@SQ    SN:chr1 LN:247249719
@SQ    SN:chr10        LN:135374737
@SQ    SN:chr11        LN:134452384
@SQ    SN:chr11_random LN:215294
@SQ    SN:chr12        LN:132349534
@SQ    SN:chr13        LN:114142980
@SQ    SN:chr13_random LN:186858
@SQ    SN:chr14        LN:106368585
@SQ    SN:chr15        LN:100338915
@SQ    SN:chr15_random LN:784346
@SQ    SN:chr16        LN:88827254
@SQ    SN:chr17        LN:78774742
@SQ    SN:chr17_random LN:2617613
@SQ    SN:chr18        LN:76117153
@SQ    SN:chr18_random LN:4262
@SQ    SN:chr19        LN:63811651
@SQ    SN:chr19_random LN:301858
@SQ    SN:chr1_random  LN:1663265
@SQ    SN:chr2 LN:242951149
@SQ    SN:chr20        LN:62435964
@SQ    SN:chr21        LN:46944323
@SQ    SN:chr21_random LN:1679693
@SQ    SN:chr22        LN:49691432
@SQ    SN:chr22_random LN:257318
@SQ    SN:chr3 LN:199501827
@SQ    SN:chr3_random  LN:749256
@SQ    SN:chr4 LN:191273063
@SQ    SN:chr4_random  LN:842648
@SQ    SN:chr5 LN:180857866
@SQ    SN:chr6 LN:170899992
@SQ    SN:chr6_random  LN:1875562
@SQ    SN:chr7 LN:158821424
@SQ    SN:chr7_random  LN:549659
@SQ    SN:chr8 LN:146274826
@SQ    SN:chr8_random  LN:943810
@SQ    SN:chr9 LN:140273252
@SQ    SN:chr9_random  LN:1146434
@SQ    SN:chrM LN:16571
@SQ    SN:chrX LN:154913754
@SQ    SN:chrX_random  LN:1719168
@SQ    SN:chrY LN:57772954
@RG    ID:Sample_618545BE      SM:Sample_618545BE      LB:Sample_618545BE      PL:Illumina
@PG    PN:bwa  ID:bwa  VN:0.6.2-r126
@PG    ID:CrossMap     VN:0.1.3
@CO    Liftover from original BAM/SAM file: test.hg19.bam
@CO    Liftover is based on the chain file: ../test/hg19ToHg18.over.chain.gz
</pre></div>
</div>
<p><strong>Optional tags:</strong></p>
<dl class="simple">
<dt>Q</dt><dd><p>QC. QC failed.</p>
</dd>
<dt>N</dt><dd><p>Unmapped. Originally unmapped or originally mapped but failed to lift over to new assembly.</p>
</dd>
<dt>M</dt><dd><p>Multiple mapped. Alignment can be lifted over to multiple places.</p>
</dd>
<dt>U</dt><dd><p>Unique mapped. Alignment can be lifted over to only 1 place.</p>
</dd>
</dl>
<p><strong>Tags for pair-end sequencing include:</strong></p>
<ul class="simple">
<li><p>QF = QC failed</p></li>
<li><p>NN = both read1 and read2 unmapped</p></li>
<li><p>NU = read1 unmapped, read2 unique mapped</p></li>
<li><p>NM = read1 unmapped, multiple mapped</p></li>
<li><p>UN = read1 uniquely mapped, read2 unmap</p></li>
<li><p>UU = both read1 and read2 uniquely mapped</p></li>
<li><p>UM = read1 uniquely mapped, read2 multiple mapped</p></li>
<li><p>MN = read1 multiple mapped, read2 unmapped</p></li>
<li><p>MU = read1 multiple mapped, read2 unique mapped</p></li>
<li><p>MM = both read1 and read2 multiple mapped</p></li>
</ul>
<p><strong>Tags for single-end sequencing include:</strong></p>
<ul class="simple">
<li><p>QF = QC failed</p></li>
<li><p>SN = unmaped</p></li>
<li><p>SM = multiple mapped</p></li>
<li><p>SU = uniquely mapped</p></li>
</ul>
<p>NOTE:</p>
<ol class="arabic simple">
<li><p>All alignments (mapped, partial mapped, unmapped, QC failed) will write to one file. Users can filter them by tags (this is why ‘-a’ is always recommended).</p></li>
<li><p>Header section will be updated to the target assembly.</p></li>
<li><p>Genome coordinates and all SAM flags in the alignment section will be updated to the target assembly.</p></li>
<li><p>if the input is a CRAM file, pysam must &gt;= 0.8.2</p></li>
<li><p>Optional fields in the alignment section will not be updated.</p></li>
</ol>
</div>
<div class="section" id="convert-wiggle-format-files">
<h2>Convert Wiggle format files<a class="headerlink" href="#convert-wiggle-format-files" title="Permalink to this headline">¶</a></h2>
<p><a class="reference external" href="http://genome.ucsc.edu/goldenPath/help/wiggle.html">Wiggle</a> (WIG) format is useful for
displaying continuous data such as GC content and the reads intensity of high-throughput sequencing data.
BigWig is a self-indexed binary-format Wiggle file, and has the advantage of supporting random access.
Input wiggle data can be in variableStep (for data with irregular intervals) or fixedStep
(for data with regular intervals). Regardless of the input, the output files are always in bedGraph
format. We export files in bedGraph format because it’s more compact than wiggle format, and more importantly,
CrossMap internally transforms wiggle into bedGraph to increase running speed.</p>
<p>Typing command without any arguments will print help message:</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span>$ python2.7 CrossMap.py  wig

Usage
-----
  CrossMap.py wig chain_file input_wig_file output_prefix

Description
-----------
  Convert wiggle format file. The &quot;chain_file&quot; can be regular or compressed (*.gz,
  *.Z, *.z, *.bz, *.bz2, *.bzip2) file, local file or URL (http://, https://,
  ftp://) pointing to remote file.  Both &quot;variableStep&quot; and &quot;fixedStep&quot; wiggle
  lines are supported. Wiggle format:
  http://genome.ucsc.edu/goldenPath/help/wiggle.html

Example
-------
  CrossMap.py wig hg18ToHg19.over.chain.gz test.hg18.wig test.hg19
</pre></div>
</div>
<p>NOTE:</p>
<ol class="arabic simple">
<li><p>To improve performance, this script calls <a class="reference external" href="http://www.gnu.org/software/coreutils/manual/html_node/sort-invocation.html">GNU “sort”</a> command internally.
If “sort” command does not exist, CrossMap will exit.</p></li>
</ol>
</div>
<div class="section" id="convert-bigwig-format-files">
<h2>Convert BigWig format files<a class="headerlink" href="#convert-bigwig-format-files" title="Permalink to this headline">¶</a></h2>
<p>If an input file is in BigWig format, the output is BigWig format if UCSC’s
‘<a class="reference external" href="http://hgdownload.cse.ucsc.edu/admin/exe/">wigToBigWig</a>’ executable can be found;
otherwise, the output file will be in bedGraph format.</p>
<p>After v0.3.0, UCSC’s wigToBigWig is no longer needed.</p>
<p>Typing command without any arguments will print help message:</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span>$ python2.7 CrossMap.py  bigwig

Usage
-----
  CrossMap.py bigwig chain_file input_bigwig_file output_prefix

Description
-----------
  Convert BigWig format file. The &quot;chain_file&quot; can be regular or compressed (*.gz,
  *.Z, *.z, *.bz, *.bz2, *.bzip2) file, local file or URL (http://, https://,
  ftp://) pointing to remote file.

Example
-------
  CrossMap.py bigwig hg18ToHg19.over.chain.gz test.hg18.bw test.hg19
</pre></div>
</div>
<p>Example (Convert BigWig file from hg18 to hg19):</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span>$ python CrossMap.py bigwig  hg19ToHg18.over.chain.gz  test.hg19.bw test.hg18
@ 2013-11-17 22:12:42: Read chain_file:  ../data/hg19ToHg18.over.chain.gz
@ 2013-11-17 22:12:44: Liftover bigwig file: test.hg19.bw ==&gt; test.hg18.bgr
@ 2013-11-17 22:15:38: Merging overlapped entries in bedGraph file ...
@ 2013-11-17 22:15:38: Sorting bedGraph file:test.hg18.bgr
@ 2013-11-17 22:15:39: Convert wiggle to bigwig ...
</pre></div>
</div>
<p>NOTE:</p>
<ol class="arabic simple">
<li><p>To improve performance, this script calls <a class="reference external" href="http://www.gnu.org/software/coreutils/manual/html_node/sort-invocation.html">GNU “sort”</a> command
internally. If “sort” command does not exist, CrossMap will exit.</p></li>
<li><p>Output files: output_prefix.bw, output_prefix.bgr, output_prefix.sorted.bgr</p></li>
</ol>
</div>
<div class="section" id="convert-gff-gtf-format-files">
<h2>Convert GFF/GTF format files<a class="headerlink" href="#convert-gff-gtf-format-files" title="Permalink to this headline">¶</a></h2>
<p><a class="reference external" href="http://genome.ucsc.edu/FAQ/FAQformat.html#format3">GFF</a> (General Feature Format) is another
plain text file used to describe gene structure. <a class="reference external" href="http://genome.ucsc.edu/FAQ/FAQformat.html#format4">GTF</a>
(Gene Transfer Format) is a refined version of GTF. The first eight fields are the same as
GFF. Plain text, compressed plain text, and URLs pointing to remote files are all supported.
Only chromosome and genome coordinates are updated. The format of the output is determined from
the input.</p>
<p>Typing command without any arguments will print help message:</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span>$ python2.7 CrossMap.py  gff

Usage
-----
  CrossMap.py gff chain_file input_gff_file output_file

Description
-----------
  Convert GFF or GTF format file. The&quot;chain_file&quot; can be regular or compressed
  (*.gz, *.Z, *.z, *.bz, *.bz2, *.bzip2) file, local file or URL (http://,
  https://, ftp://) pointing to remote file. Input file must be in GFF or GTF
  format. GFF format: http://genome.ucsc.edu/FAQ/FAQformat.html#format3 GTF
  format: http://genome.ucsc.edu/FAQ/FAQformat.html#format4

Example1 (write output to file)
-------------------------------
  CrossMap.py gff  hg19ToHg18.over.chain.gz test.hg19.gtf test.hg18.gtf

Example2 (write output to screen)
---------------------------------
  CrossMap.py gff  hg19ToHg18.over.chain.gz test.hg19.gtf
</pre></div>
</div>
<p>Example (Convert GTF file from hg19 to hg18):</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span>$ python CrossMap.py gff  hg19ToHg18.over.chain.gz test.hg19.gtf test.hg18.gtf
@ 2013-11-17 20:44:47: Read chain_file:  ../data/hg19ToHg18.over.chain.gz

$ head test.hg19.gtf
chr1   hg19_refGene    CDS     48267145        48267291        0.000000        -       0       gene_id &quot;NM_001194986&quot;; transcript_id &quot;NM_001194986&quot;;
chr1   hg19_refGene    exon    66081691        66081907        0.000000        +       .       gene_id &quot;NM_002303&quot;; transcript_id &quot;NM_002303&quot;;
chr1   hg19_refGene    CDS     145334684       145334792       0.000000        +       2       gene_id &quot;NM_001039703&quot;; transcript_id &quot;NM_001039703&quot;;
chr1   hg19_refGene    exon    172017752       172017890       0.000000        +       .       gene_id &quot;NM_001136127&quot;; transcript_id &quot;NM_001136127&quot;;
chr1   hg19_refGene    CDS     206589249       206589333       0.000000        +       2       gene_id &quot;NM_001170637&quot;; transcript_id &quot;NM_001170637&quot;;
chr1   hg19_refGene    exon    210573812       210574006       0.000000        +       .       gene_id &quot;NM_001170580&quot;; transcript_id &quot;NM_001170580&quot;;
chr1   hg19_refGene    CDS     235850249       235850347       0.000000        -       0       gene_id &quot;NM_000081&quot;; transcript_id &quot;NM_000081&quot;;
chr1   hg19_refGene    CDS     235880012       235880078       0.000000        -       1       gene_id &quot;NM_000081&quot;; transcript_id &quot;NM_000081&quot;;
chr1   hg19_refGene    exon    3417741 3417872 0.000000        -       .       gene_id &quot;NM_001409&quot;; transcript_id &quot;NM_001409&quot;;
chr1   hg19_refGene    exon    10190773        10190871        0.000000        +       .       gene_id &quot;NM_006048&quot;; transcript_id &quot;NM_006048&quot;;

$ head test.hg18.gtf
chr1   hg19_refGene    CDS     48039732        48039878        0.000000        -       0       gene_id &quot;NM_001194986&quot;; transcript_id &quot;NM_001194986&quot;;
chr1   hg19_refGene    exon    65854279        65854495        0.000000        +       .       gene_id &quot;NM_002303&quot;; transcript_id &quot;NM_002303&quot;;
chr1   hg19_refGene    CDS     144046041       144046149       0.000000        +       2       gene_id &quot;NM_001039703&quot;; transcript_id &quot;NM_001039703&quot;;
chr1   hg19_refGene    exon    170284375       170284513       0.000000        +       .       gene_id &quot;NM_001136127&quot;; transcript_id &quot;NM_001136127&quot;;
chr1   hg19_refGene    CDS     204655872       204655956       0.000000        +       2       gene_id &quot;NM_001170637&quot;; transcript_id &quot;NM_001170637&quot;;
chr1   hg19_refGene    exon    208640435       208640629       0.000000        +       .       gene_id &quot;NM_001170580&quot;; transcript_id &quot;NM_001170580&quot;;
chr1   hg19_refGene    CDS     233916872       233916970       0.000000        -       0       gene_id &quot;NM_000081&quot;; transcript_id &quot;NM_000081&quot;;
chr1   hg19_refGene    CDS     233946635       233946701       0.000000        -       1       gene_id &quot;NM_000081&quot;; transcript_id &quot;NM_000081&quot;;
chr1   hg19_refGene    exon    3407601 3407732 0.000000        -       .       gene_id &quot;NM_001409&quot;; transcript_id &quot;NM_001409&quot;;
chr1   hg19_refGene    exon    10113360        10113458        0.000000        +       .       gene_id &quot;NM_006048&quot;; transcript_id &quot;NM_006048&quot;;
</pre></div>
</div>
<p>NOTE:</p>
<ol class="arabic simple">
<li><p>Each feature  (exon, intron, UTR, etc) is processed separately and independently, and
we do NOT check if features originally belonging to the same gene were converted into the same gene.</p></li>
<li><p>If user wants to lift over gene annotation files, use BED12 format.</p></li>
<li><p>If no output file was specified, the output will be printed to screen (console). In this case, items failed to convert are also printed out.</p></li>
</ol>
</div>
<div class="section" id="convert-vcf-format-files">
<h2>Convert VCF format files<a class="headerlink" href="#convert-vcf-format-files" title="Permalink to this headline">¶</a></h2>
<p><a class="reference external" href="http://www.1000genomes.org/wiki/Analysis/Variant%20Call%20Format/vcf-variant-call-format-version-41">VCF</a>
(variant call format) is a flexible and extendable line-oriented text format developed by
the <a class="reference external" href="http://www.1000genomes.org/">1000 Genome Project</a>. It is useful for representing single
nucleotide variants, indels, copy number variants, and structural variants. Chromosomes,
coordinates, and reference alleles are updated to a new assembly, and all the other fields
are not changed.</p>
<p>Typing command without any arguments will print help message:</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span>$ python2.7 CrossMap.py  vcf

usage
-----
  CrossMap.py vcf chain_file input_VCF_file ref_genome_file output_file

Description
-----------
  Convert VCF format file. The &quot;chain_file&quot; and &quot;input_VCF_file&quot; can be regular or
  compressed (*.gz, *.Z, *.z, *.bz, *.bz2, *.bzip2) file, local file or URL
  (http://, https://, ftp://) pointing to remote file. &quot;ref_genome_file&quot; is genome
  sequence file of &#39;target assembly&#39; in FASTA format.

Example
-------
   CrossMap.py vcf hg19ToHg18.over.chain.gz test.hg19.vcf hg18.fa test.hg18.vcf
</pre></div>
</div>
<p>Example (Convert VCF file from hg19 to hg18):</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span>$ python CrossMap.py vcf hg19ToHg18.over.chain.gz test.hg19.vcf ../database/genome/hg18.fa  test.hg18.vcf
@ 2015-07-27 10:14:23: Read chain_file:  ../data/hg19ToHg18.over.chain.gz
@ 2013-11-17 20:53:39: Creating index for ../database/genome/hg18.fa
@ 2015-07-27 10:14:50: Total entries: 497
@ 2015-07-27 10:14:50: Failed to map: 0

$ grep -v &#39;#&#39; test.hg19.vcf  |head -10
chr1   10933566        .       C       G       .       PASS    ADP=13;WT=0;HET=0;HOM=1;NC=0    GT:GQ:SDP:DP:RD:AD:FREQ:PVAL:RBQ:ABQ:RDF:RDR:ADF:ADR    1/1:7:13:13:0:13:100%:9.6148E-8:0:36:0:0:8:5
chr1   11187893        .       T       C       .       PASS    ADP=224;WT=0;HET=0;HOM=1;NC=0   GT:GQ:SDP:DP:RD:AD:FREQ:PVAL:RBQ:ABQ:RDF:RDR:ADF:ADR    1/1:133:226:224:0:224:100%:3.6518E-134:0:38:0:0:41:183
chr1   11205058        .       C       T       .       PASS    ADP=625;WT=0;HET=0;HOM=1;NC=0   GT:GQ:SDP:DP:RD:AD:FREQ:PVAL:RBQ:ABQ:RDF:RDR:ADF:ADR    1/1:255:643:625:0:625:100%:0E0:0:37:0:0:294:331
chr1   11292753        .       A       G       .       PASS    ADP=52;WT=0;HET=0;HOM=1;NC=0    GT:GQ:SDP:DP:RD:AD:FREQ:PVAL:RBQ:ABQ:RDF:RDR:ADF:ADR    1/1:27:52:52:2:50:96.15%:9.0394E-28:39:38:0:2:0:50
chr1   11318763        .       C       G       .       str10   ADP=88;WT=0;HET=0;HOM=1;NC=0    GT:GQ:SDP:DP:RD:AD:FREQ:PVAL:RBQ:ABQ:RDF:RDR:ADF:ADR    1/1:51:88:88:0:88:100%:1.7384E-52:0:38:0:0:1:87
chr1   11319587        .       A       G       .       PASS    ADP=70;WT=0;HET=0;HOM=1;NC=0    GT:GQ:SDP:DP:RD:AD:FREQ:PVAL:RBQ:ABQ:RDF:RDR:ADF:ADR    1/1:40:70:70:0:70:100%:1.0659E-41:0:38:0:0:0:70
chr1   16202995        .       C       T       .       PASS    ADP=463;WT=0;HET=1;HOM=0;NC=0   GT:GQ:SDP:DP:RD:AD:FREQ:PVAL:RBQ:ABQ:RDF:RDR:ADF:ADR    0/1:1:463:463:458:5:1.08%:3.0913E-2:37:33:188:270:4:1
chr1   27088546        .       A       T       .       PASS    ADP=124;WT=0;HET=1;HOM=0;NC=0   GT:GQ:SDP:DP:RD:AD:FREQ:PVAL:RBQ:ABQ:RDF:RDR:ADF:ADR    0/1:21:124:124:65:59:47.58%:1.7915E-22:37:38:59:6:55:4
chr1   27101390        .       T       C       .       str10   ADP=267;WT=0;HET=1;HOM=0;NC=0   GT:GQ:SDP:DP:RD:AD:FREQ:PVAL:RBQ:ABQ:RDF:RDR:ADF:ADR    0/1:1:267:267:262:5:1.87%:3.0665E-2:32:22:85:177:5:0
chr1   34007097        .       T       C       .       PASS    ADP=10;WT=0;HET=1;HOM=0;NC=0    GT:GQ:SDP:DP:RD:AD:FREQ:PVAL:RBQ:ABQ:RDF:RDR:ADF:ADR    0/1:1:10:10:6:4:40%:4.3344E-2:34:32:0:6:0:4

$ grep -v &#39;#&#39; test.hg18.vcf  |head -10
1      10856153        .       C       G       .       PASS    ADP=13;WT=0;HET=0;HOM=1;NC=0    GT:GQ:SDP:DP:RD:AD:FREQ:PVAL:RBQ:ABQ:RDF:RDR:ADF:ADR    1/1:7:13:13:0:13:100%:9.6148E-8:0:36:0:0:8:5
1      11110480        .       T       C       .       PASS    ADP=224;WT=0;HET=0;HOM=1;NC=0   GT:GQ:SDP:DP:RD:AD:FREQ:PVAL:RBQ:ABQ:RDF:RDR:ADF:ADR    1/1:133:226:224:0:224:100%:3.6518E-134:0:38:0:0:41:183
1      11127645        .       C       T       .       PASS    ADP=625;WT=0;HET=0;HOM=1;NC=0   GT:GQ:SDP:DP:RD:AD:FREQ:PVAL:RBQ:ABQ:RDF:RDR:ADF:ADR    1/1:255:643:625:0:625:100%:0E0:0:37:0:0:294:331
1      11215340        .       A       G       .       PASS    ADP=52;WT=0;HET=0;HOM=1;NC=0    GT:GQ:SDP:DP:RD:AD:FREQ:PVAL:RBQ:ABQ:RDF:RDR:ADF:ADR    1/1:27:52:52:2:50:96.15%:9.0394E-28:39:38:0:2:0:50
1      11241350        .       C       G       .       str10   ADP=88;WT=0;HET=0;HOM=1;NC=0    GT:GQ:SDP:DP:RD:AD:FREQ:PVAL:RBQ:ABQ:RDF:RDR:ADF:ADR    1/1:51:88:88:0:88:100%:1.7384E-52:0:38:0:0:1:87
1      11242174        .       A       G       .       PASS    ADP=70;WT=0;HET=0;HOM=1;NC=0    GT:GQ:SDP:DP:RD:AD:FREQ:PVAL:RBQ:ABQ:RDF:RDR:ADF:ADR    1/1:40:70:70:0:70:100%:1.0659E-41:0:38:0:0:0:70
1      16075582        .       C       T       .       PASS    ADP=463;WT=0;HET=1;HOM=0;NC=0   GT:GQ:SDP:DP:RD:AD:FREQ:PVAL:RBQ:ABQ:RDF:RDR:ADF:ADR    0/1:1:463:463:458:5:1.08%:3.0913E-2:37:33:188:270:4:1
1      26961133        .       A       T       .       PASS    ADP=124;WT=0;HET=1;HOM=0;NC=0   GT:GQ:SDP:DP:RD:AD:FREQ:PVAL:RBQ:ABQ:RDF:RDR:ADF:ADR    0/1:21:124:124:65:59:47.58%:1.7915E-22:37:38:59:6:55:4
1      26973977        .       T       C       .       str10   ADP=267;WT=0;HET=1;HOM=0;NC=0   GT:GQ:SDP:DP:RD:AD:FREQ:PVAL:RBQ:ABQ:RDF:RDR:ADF:ADR    0/1:1:267:267:262:5:1.87%:3.0665E-2:32:22:85:177:5:0
1      33779684        .       T       C       .       PASS    ADP=10;WT=0;HET=1;HOM=0;NC=0    GT:GQ:SDP:DP:RD:AD:FREQ:PVAL:RBQ:ABQ:RDF:RDR:ADF:ADR    0/1:1:10:10:6:4:40%:4.3344E-2:34:32:0:6:0:4

$ grep -v &#39;#&#39; test.hg18.vcf.unmap      #coordinates are still based on hg19
chr14  20084444        .       G       C       .       PASS    ADP=253;WT=0;HET=1;HOM=0;NC=0   GT:GQ:SDP:DP:RD:AD:FREQ:PVAL:RBQ:ABQ:RDF:RDR:ADF:ADR    0/1:1:253:253:247:5:1.98%:3.0631E-2:38:39:123:124:5:0
chr14  20086290        .       T       C       .       PASS    ADP=441;WT=0;HET=1;HOM=0;NC=0   GT:GQ:SDP:DP:RD:AD:FREQ:PVAL:RBQ:ABQ:RDF:RDR:ADF:ADR    0/1:4:441:441:427:14:3.17%:5.4963E-5:37:38:236:191:6:8
</pre></div>
</div>
<p>NOTE:</p>
<ol class="arabic simple">
<li><p>Genome coordinates and reference allele will be updated to target assembly.</p></li>
<li><p>Reference genome is genome sequence of target assembly.</p></li>
<li><p>If the reference genome sequence file (../database/genome/hg18.fa) was not indexed, CrossMap will automatically indexed it (only the first time you run CrossMap).</p></li>
<li><p>Output files: <em>output_file</em> and <em>output_file.unmap</em>.</p></li>
<li><p>In the output VCF file, whether the chromosome IDs contain “chr” or not depends on the format of the input VCF file.</p></li>
</ol>
</div>
<div class="section" id="convert-maf-format-files">
<h2>Convert MAF format files<a class="headerlink" href="#convert-maf-format-files" title="Permalink to this headline">¶</a></h2>
<p><a class="reference external" href="https://docs.gdc.cancer.gov/Encyclopedia/pages/Mutation_Annotation_Format/">MAF</a>
(mutation annotation format) files are tab-delimited
files that contain somatic and/or germline mutation annotations. Please do not confused
with <a class="reference external" href="https://genome.ucsc.edu/FAQ/FAQformat.html#format5">Multiple Alignment Format</a>.</p>
<p>Typing command without any arguments will print help message:</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span>$ python2.7 CrossMap.py  maf

usage
-----
 CrossMap.py maf chain_file input_MAF_file ref_genome_file build_name output_file

Description
-----------
 Convert MAF format file. The &quot;chain_file&quot; and &quot;input_MAF_file&quot; can be regular or
 compressed (*.gz, *.Z, *.z, *.bz, *.bz2, *.bzip2) file, local file or URL
 (http://, https://, ftp://) pointing to remote file. &quot;ref_genome_file&quot; is genome
 sequence file of &#39;target assembly&#39; in FASTA format. &quot;build_name&quot; is the name of
 the &#39;target_assembly&#39; (eg &quot;GRCh38&quot;)

Example
-------
  CrossMap.py  maf  hg19ToHg38.over.chain.gz  test.hg19.maf  hg38.fa  GRCh38  test.hg38.maf
</pre></div>
</div>
</div>
</div>
<div class="section" id="compare-to-ucsc-liftover-tool">
<h1>Compare to UCSC liftover tool<a class="headerlink" href="#compare-to-ucsc-liftover-tool" title="Permalink to this headline">¶</a></h1>
<p>To access the accuracy of CrossMap, we randomly generated 10,000 genome intervals (download from <a class="reference external" href="https://sourceforge.net/projects/crossmap/files/hg19.rand.bed.gz/download">here</a>) with the
fixed interval size of 200 bp from hg19. Then we converted them into hg18 using CrossMap
and <a class="reference external" href="http://genome.ucsc.edu/cgi-bin/hgLiftOver">UCSC liftover tool</a> with default configurations. We compare CrossMap
to <a class="reference external" href="http://genome.ucsc.edu/cgi-bin/hgLiftOver">UCSC liftover tool</a> because it is the most widely
used tool to convert genome coordinates.</p>
<p>CrossMap failed to convert 613 intervals, and UCSC liftover tool failed to convert 614
intervals. All failed intervals are exactly the same except one region (chr2 90542908 90543108).
UCSC failed to convert it because this region needs to be split twice:</p>
<table class="docutils align-default">
<colgroup>
<col style="width: 29%" />
<col style="width: 30%" />
<col style="width: 40%" />
</colgroup>
<thead>
<tr class="row-odd"><th class="head"><p>Original (hg19)</p></th>
<th class="head"><p>Split (hg19)</p></th>
<th class="head"><p>Target (hg18)</p></th>
</tr>
</thead>
<tbody>
<tr class="row-even"><td><p>chr2 90542908  90543108 -</p></td>
<td><p>chr2 90542908 90542933 -</p></td>
<td><p>chr2    89906445        89906470 -</p></td>
</tr>
<tr class="row-odd"><td><p>chr2 90542908  90543108 -</p></td>
<td><p>chr2 90542933 90543001 -</p></td>
<td><p>chr2    87414583        87414651 -</p></td>
</tr>
<tr class="row-even"><td><p>chr2 90542908  90543108 -</p></td>
<td><p>chr2 90543010 90543108 -</p></td>
<td><p>chr2    87414276        87414374 -</p></td>
</tr>
</tbody>
</table>
<p>For genome intervals that were successfully converted to hg18, the start and end coordinates are
exactly the same between UCSC conversion and CrossMap conversion.</p>
<a class="reference internal image-reference" href="backup/_static/CrossMap_vs_UCSC.png"><img alt="CrossMap_vs_UCSC_liftover.png" src="backup/_static/CrossMap_vs_UCSC.png" style="width: 700.0px; height: 400.0px;" /></a>
</div>
<div class="section" id="citation">
<h1>Citation<a class="headerlink" href="#citation" title="Permalink to this headline">¶</a></h1>
<p>Zhao, H., Sun, Z., Wang, J., Huang, H., Kocher, J.-P., &amp; Wang, L. (2013). CrossMap: a versatile tool for coordinate conversion between genome assemblies. Bioinformatics (Oxford, England), btt730.</p>
</div>
<div class="section" id="license">
<h1>LICENSE<a class="headerlink" href="#license" title="Permalink to this headline">¶</a></h1>
<p>CrossMap is distributed under <a class="reference external" href="http://www.gnu.org/copyleft/gpl.html">GNU General Public License</a></p>
<p>This program is free software; you can redistribute it and/or
modify it under the terms of the GNU General Public License as
published by the Free Software Foundation; either version 2 of the
License, or (at your option) any later version. This program is distributed in the hope that it will be useful,
but WITHOUT ANY WARRANTY; without even the implied warranty of
MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
General Public License for more details. You should have received a copy of the GNU General Public License
along with this program; if not, write to the Free Software
Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA
02110-1301 USA</p>
</div>
<div class="section" id="contact">
<h1>Contact<a class="headerlink" href="#contact" title="Permalink to this headline">¶</a></h1>
<ul class="simple">
<li><p>Wang.Liguo AT mayo.edu</p></li>
</ul>
</div>


          </div>
        </div>
      </div>
      <div class="sphinxsidebar" role="navigation" aria-label="main navigation">
        <div class="sphinxsidebarwrapper">
  <h3><a href="../index.html">Table of Contents</a></h3>
  <ul>
<li><a class="reference internal" href="#">What is CrossMap ?</a></li>
<li><a class="reference internal" href="#how-crossmap-works">How CrossMap works?</a></li>
<li><a class="reference internal" href="#release-history">Release history</a></li>
<li><a class="reference internal" href="#installation">Installation</a><ul>
<li><a class="reference internal" href="#use-pip-to-install-crossmap">Use pip to install CrossMap</a></li>
<li><a class="reference internal" href="#use-pip-to-upgrade-crossmap">Use pip to upgrade CrossMap</a></li>
</ul>
</li>
<li><a class="reference internal" href="#input-and-output">Input and Output</a><ul>
<li><a class="reference internal" href="#chain-file">Chain file</a></li>
<li><a class="reference internal" href="#user-input-file">User Input file</a></li>
<li><a class="reference internal" href="#output-file">Output file</a></li>
</ul>
</li>
<li><a class="reference internal" href="#usage">Usage</a><ul>
<li><a class="reference internal" href="#convert-bed-format-files">Convert BED format files</a></li>
<li><a class="reference internal" href="#convert-bam-cram-sam-format-files">Convert BAM/CRAM/SAM format files</a></li>
<li><a class="reference internal" href="#convert-wiggle-format-files">Convert Wiggle format files</a></li>
<li><a class="reference internal" href="#convert-bigwig-format-files">Convert BigWig format files</a></li>
<li><a class="reference internal" href="#convert-gff-gtf-format-files">Convert GFF/GTF format files</a></li>
<li><a class="reference internal" href="#convert-vcf-format-files">Convert VCF format files</a></li>
<li><a class="reference internal" href="#convert-maf-format-files">Convert MAF format files</a></li>
</ul>
</li>
<li><a class="reference internal" href="#compare-to-ucsc-liftover-tool">Compare to UCSC liftover tool</a></li>
<li><a class="reference internal" href="#citation">Citation</a></li>
<li><a class="reference internal" href="#license">LICENSE</a></li>
<li><a class="reference internal" href="#contact">Contact</a></li>
</ul>

  <div role="note" aria-label="source link">
    <h3>This Page</h3>
    <ul class="this-page-menu">
      <li><a href="../_sources/backup/index.rst.txt"
            rel="nofollow">Show Source</a></li>
    </ul>
   </div>
<div id="searchbox" style="display: none" role="search">
  <h3 id="searchlabel">Quick search</h3>
    <div class="searchformwrapper">
    <form class="search" action="../search.html" method="get">
      <input type="text" name="q" aria-labelledby="searchlabel" />
      <input type="submit" value="Go" />
    </form>
    </div>
</div>
<script type="text/javascript">$('#searchbox').show(0);</script>
        </div>
      </div>
      <div class="clearer"></div>
    </div>
    <div class="related" role="navigation" aria-label="related navigation">
      <h3>Navigation</h3>
      <ul>
        <li class="right" style="margin-right: 10px">
          <a href="../genindex.html" title="General Index"
             >index</a></li>
        <li class="nav-item nav-item-0"><a href="../index.html">CrossMap 0.4.0 documentation</a> &#187;</li> 
      </ul>
    </div>
    <div class="footer" role="contentinfo">
        &#169; Copyright 2019, Liguo Wang.
      Created using <a href="http://sphinx-doc.org/">Sphinx</a> 2.1.2.
    </div>
  </body>
</html>